```{r setup}
library(magrittr)
library(ggplot2)
library(GenomicRanges)

library(org.Dm.eg.db)
library(AnnotationDbi)

library(EnrichedHeatmap)
library(ComplexHeatmap)

source('utils.R')

###
# load rData
###

load('rData/sheets.rda')
load('rData/peaks.rda')

###
# assign global variables
###

dm6 <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6
dm6.TxDb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene::TxDb.Dmelanogaster.UCSC.dm6.ensGene

brD <- data.frame('seqnames' = 'chrX',
                  'start' = 1565708,
                  'end' = 1567401) %>%
  GenomicRanges::GRanges()

###
# colors
### 

viridis.hex <- c("#440154","#3b528b","#21918c","#5ec962","#5ec962","#fde725")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


###
# peak subsets
###

osaPeaks <- peaks %>% 
  dplyr::filter(osa.cnr & !yw.cnr & assay == 'cnr') %>% 
  dplyr::mutate(deg.peakCat = dplyr::case_when(osaGFP.deGrad & !osaGFP.control ~ 'osa ectopic',
                                               !osaGFP.deGrad & osaGFP.control ~ 'osa dependent',
                                               osaGFP.deGrad & osaGFP.control ~ 'osa independent',
                                               T ~ as.character(NA))) %>%  # NA here would be peaks that aren't reproducible or of good quality
  GRanges() %>% 
  resize(width = 1, fix = "center")

fairePeaks.wt <- peaks %>%
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(width = 1, fix = "center")

fairePeaks.deg <- peaks %>%
  dplyr::filter(assay == 'faire' & experiment == 'osaGFP deGrad Pupal Wing FAIRE') %>%
  dplyr::mutate(deg.peakCat = dplyr::case_when(osaGFP.deGrad & !osaGFP.control ~ 'osa ectopic',
                                               !osaGFP.deGrad & osaGFP.control ~ 'osa dependent',
                                               osaGFP.deGrad & osaGFP.control ~ 'osa independent',
                                               T ~ as.character(NA))) %>%  # NA here would be peaks that aren't reproducible or of good quality
  dplyr::filter(!is.na(deg.peakCat)) %>%
  GRanges() %>%
  resize(width = 1, fix = "center")

###
# load bws and matrices for heatmaps
### 

#TODO - cleanup - move matrices?

bws <- get_bws(dplyr::bind_rows(cnr.ss, faire.wt.ssPool), 'id') 

deg.bws <- get_bws(faire.ss %>% dplyr::filter(experiment == 'osaGFP deGrad FAIRE'), 'id')

mats <- purrr::map(bws, ~{
  normalizeToMatrix(., osaPeaks, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
})
names(mats) <- names(bws)

```

TODO - make reference table of different peak categories and parameters, eg. how osa specific peaks are defined, how closing peaks are defined
TODO - currently defining dynamic faire simply by log2fold change - no peak call requirement
# Fig 1

## Fig 1A. - brDisc browser shot
```{r, 1A}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("grey20")

faire.wt.browser <- faire.wt.ssPool %>%
  dplyr::filter(grepl('3LW',time)) %>%
  dplyr::mutate(color = track.colors)

faire.wt.tracks <- get_cnr_tracks(faire.wt.browser, ylim = c(0,15))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = '#e858a5', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)

genetrack <- Gviz::GeneRegionTrack(range = dm6.TxDb, fill = 'grey80', showTitle = F, background.title = 'white')

png('rPlots/1A.png', width = 10, height = 2, res = 300, units = 'in')
Gviz::plotTracks(c(axis,faire.wt.tracks, brdisc.at, genetrack), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 20000),
                 to = end(brD + 20000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```


# Fig 3


## Fig 3A. heatmap of osa binding sites vs control

```{r, 3A}

#TODO - functionalize normalizetomatrix call, + common min/max and colorRamp

cnr.mats <- list(mats$yw.rep1, mats$osaGFP.rep1, mats$yw.rep2, mats$osaGFP.rep2) # correct order for subsequent names()
names(cnr.mats) <- cnr.ss$id

common_min = min(unlist(cnr.mats))
common_max = max(unlist(cnr.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

osa.count <- osaPeaks %>%
  data.frame() %>%
  nrow()

osa.count[1] <- as.character(osa.count[1])
  
names(osa.count) <- 'Osa Peak'

groups <- ifelse(osaPeaks$osa.cnr, 'Osa Peak', NA)

cnr.hms <- purrr::map(names(cnr.mats), function(x) {
  if(x == 'osaGFP.rep1'){
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,6))),
                    left_annotation = c(rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(osa.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm'))))) 
  } else {
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun, 
                    show_heatmap_legend = FALSE,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis = F, ylim = c(0,6)))) 
  }
})
names(cnr.hms) <- names(cnr.mats)

hms.ordered <-cnr.hms$osaGFP.rep1 + cnr.hms$osaGFP.rep2 + cnr.hms$yw.rep1 + cnr.hms$yw.rep2

png('rPlots/3A.png', width = 6, height = 6, units = 'in', res = 300)
hms.ordered
dev.off()
```

## Fig 3B. Osa CnR brDisc browswer shot

```{r, fig.width=4, fig.height=3}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("#B8E186","#F1B6DA", "#4DAC26","#D01C8B")

cnr.ss.browser <- cnr.ss %>%
  dplyr::mutate(color = track.colors)

cnr.tracks <- get_cnr_tracks(cnr.ss.browser, ylim = c(0,12))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/3B.png', width = 4, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, cnr.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```

## Fig 3C. Where are osaGFP binding sites?
  
```{r, 3C-peak-location}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr | assay == '1kb background') %>%
  dplyr::group_by(assay, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
  ggplot(aes(x = assay, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.3, 'cm'),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/3C.png', width = 4, height = 4, units = 'in', dpi = 300)

```

## Fig 3D. How does osa binding correlate with wt faire data?

```{r 3LW-FAIRE-in-OsaPeaks}


fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks %>%
  data.frame() %>%
  dplyr::group_by(osa.cnr, WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm'))),
                                                 rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'right'))))
  

png('rPlots/3D.png', width = 5, heigh = 5, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, merge_legend = T)
dev.off()


```


## Fig 3E. Fraction of Osa binding sites withing WT 3LW-24h FAIRE categories
```{r, 3E}
#TODO - combine fraction barplots and heatmaps into single figure - cowplots? grid?

faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.wt, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.wt, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat, faire.24h.mat))
common_max = max(c(faire.3lw.mat, faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::count() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')

groups <- fairePeaks.wt$faireCat.3LW_24h

lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                      left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 

plot.frac <- function(x) {
 lapply(unique(x$change), function(i) {
   x %>%
     dplyr::filter(change == i) %>%
     ggplot(aes(fraction, change, fill = osa.cnr)) +
     geom_bar(stat = 'identity', width = 1) +
     geom_text(aes(label = per), position = position_fill(vjust = 0.6),  size = 7, fontface = "bold") +
     #geom_text(aes(label = per, position = position_stack(vjust = 0.5))) +
     scale_fill_manual(values = c('grey80','#41f0ad'), name = 'Osa Bound') +
     scale_y_discrete(position = 'right', ) +
     xlab('Percent Bound by Osa') +
     theme(legend.position = 'top',
           legend.key.size = unit(1, 'cm'),
           legend.text = element_text(size = 15),
           legend.title = element_text(size = 20), 
           panel.background = element_blank(),
           axis.text = element_blank(),
           axis.title = element_blank(),
           axis.ticks = element_blank(),
#           axis.text.y = element_text(size = 12),
           panel.grid = element_blank()) + 
     NULL
   fn = paste0('rPlots/3E_',i,'.png')
   ggsave(fn, width = 6, height = 1.5, units = 'in', dpi = 300 )
   })
}

png('rPlots/3E.png', width = 4, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hm, annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'left', )
dev.off() 
 
peaks %>% 
  dplyr::filter(assay == 'faire' & !yw.cnr & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::group_by(faireCat.3LW_24h, osa.cnr) %>%
  dplyr::count() %>%
  na.omit() %>%
  reshape2::melt(id = c("osa.cnr","n"), variable.name = 'stage', value.name = 'change') %>%
  dplyr::group_by(stage, change, osa.cnr) %>%
  dplyr::summarise(peaks = sum(n)) %>%
  dplyr::mutate(fraction = peaks / sum(peaks),
                change = factor(change, levels = c('static','opening','closing'))) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()

```

##Fig3 - Plot Drafts

```{r, 3draft-1.png}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr & faireCat.3LW_24h != 'NA') %>%
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
  ggplot(aes(x = faireCat.3LW_24h, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.01, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1,
                  ) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.3,"cm"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/drafts/3draft-1.png', width = 4, height = 4, dpi = 300)

  
```


```{r, 3draft-2.png}

peaks %>%
  dplyr::filter(assay == 'faire' & faireCat.3LW_24h != 'NA' & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::mutate(osa.specific = ifelse(osa.cnr & !yw.cnr, T, F)) %>%
 # dplyr::filter(grepl('Distal|Exon|Intron|Promoter', anno.new)) %>%
 # dplyr::mutate(anno.temp = dplyr::case_when(grepl('Distal|Intron', anno.new) ~ 'Distal or Intronic',
 #                                           grepl('Exon', anno.new) ~ 'Exon',
 #                                           grepl('Promoter', anno.new) ~ 'Promoter',
 #                                           T~anno.new)) %>%
  dplyr::group_by(osa.specific, anno.new, faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n),
                pct = round(pct, digits = 2)) %>%
  ggplot(aes(x = n, y = anno.new, fill = osa.specific)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text(aes(label = n), position = position_fill(vjust = 0.5), color = 'black') +
  scale_fill_manual(values = c('#D53B59','#52CA3E','#1E82E0')) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        axis.text.x = element_text(size = 5),
        legend.key.size = unit(0.3,"cm")) +
  facet_wrap(~faireCat.3LW_24h)

ggsave('rPlots/drafts/3draft-2.png', width = 6, height = 3, dpi = 300)
```


# Fig 4 - OsaGFP DeGrad FAIRE

There is very little change between control and osa deGrad conditions. 
Figure 4 should demonstrate this and highligh if there are any subtle changes, where they occur, and how they relate to osa binding. 


## Fig 4B. FAIRE control vs osaDeGrad - split by degrad categories
Using deseq differential scoring between control and osaGFP deGrad to define FAIRE peak categories.
Using a very sensitive cutoff for log2fold change. 
"ectopic" (gain of signal in degrad) <= log2(control/degrad) <= -0.4, 
"dependent" (loss of signal in degrad) >= log2(control/degrad) >= 0.4,
"independet" (no significant change).


```{r, 4B}

#TODO - are heatmaps misleading? is this difficult to parse when split by these different peak call groups?



deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::group_by(faireCat.osaDeGrad) %>%
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')

groups <- fairePeaks.deg$faireCat.osaDeGrad

lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4B.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'bottom', )
dev.off()

fairePeaks.deg %>% 
  data.frame() %>%
  dplyr::filter(!yw.cnr) %>%
  dplyr::group_by(faireCat.osaDeGrad, osa.cnr) %>%
  dplyr::tally() %>%
  dplyr::mutate(fraction = n / sum(n),
                change = faireCat.osaDeGrad) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()


fairePeaks.deg %>% 
  data.frame() %>%
  dplyr::filter(!yw.cnr) %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & osa.cnr)
```
## Fig 4B VARIANT. osa degrad vs control heatmap
Changes are defined by reproducible peak calls...
"ectopic" peaks are unique to osaDegrad condition, vs "dependent" peaks are unique to control condition.
This approach is perhaps misleading and more difficult to articulate.

```{r, 4B}

#TODO - are heatmaps misleading? is this difficult to parse when split by these different peak call groups?



deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::group_by(deg.peakCat) %>%
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('osa dependent', 'osa ectopic', 'osa independent')

groups <- fairePeaks.deg$deg.peakCat

lgd = Legend(at = c("osa dependent", "osa ectopic", "osa independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4B.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'bottom', )
dev.off()


fairePeaks.deg %>% 
  data.frame() %>%
  dplyr::filter(!yw.cnr) %>%
  dplyr::group_by(deg.peakCat, osa.cnr) %>%
  dplyr::tally() 
  dplyr::mutate(fraction = n / sum(n),
                change = deg.peakCat) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()

```
## Fig 4C. 
```{r}
osaClosing <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  GRanges()

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., osaClosing, value_column = 'score', keep = c(0.11,0.99), extend = 2000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

#deg.count <- fairePeaks.deg %>%
#  data.frame() %>%
#  dplyr::group_by(deg.peakCat) %>%
#  dplyr::tally() %>%
#  .$n
#
#names(deg.count) <- c('osa dependent', 'osa ectopic', 'osa independent')

#groups <- osaPeaks$deg.peakCat

#lgd = Legend(at = c("osa dependent", "osa ectopic", "osa independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,1))))
#                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
#                                                                             as.list(deg.count),
#                                                                             side = 'left',
#                                                                             background_gp = gpar(fill = NA, col = NA),
#                                                                             gp = gpar(col = 'black'), 
#                                                                             padding = unit(0, 'mm')),
#                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)))
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(ylim = c(0,1), axis = F)))
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4C.png', width = 8, height = 3, units = 'in', res = 300)
ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```
## draft
```{r}
osa.dependent <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(deg.peakCat == 'osa dependent') %>%
  GRanges()

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., osa.dependent, value_column = 'score', keep = c(0.11,0.99), extend = 2000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

#deg.count <- fairePeaks.deg %>%
#  data.frame() %>%
#  dplyr::group_by(deg.peakCat) %>%
#  dplyr::tally() %>%
#  .$n
#
#names(deg.count) <- c('osa dependent', 'osa ectopic', 'osa independent')

#groups <- osaPeaks$deg.peakCat

#lgd = Legend(at = c("osa dependent", "osa ectopic", "osa independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,5))))
#                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
#                                                                             as.list(deg.count),
#                                                                             side = 'left',
#                                                                             background_gp = gpar(fill = NA, col = NA),
#                                                                             gp = gpar(col = 'black'), 
#                                                                             padding = unit(0, 'mm')),
#                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)))
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(ylim = c(0,5), axis = F)))
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

#png('rPlots/4C.png', width = 8, height = 3, units = 'in', res = 300)
ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
#dev.off()
```


## draft 
```{r}
osa.ectopic <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(deg.peakCat == 'osa ectopic') %>%
  GRanges()

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., osa.ectopic, value_column = 'score', keep = c(0.11,0.99), extend = 2000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- osa.ectopic %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('closing', 'opening', 'static')

groups <- osa.ectopic$faireCat.3LW_24h

lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,10))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)))
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,10), axis = F)))
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

#png('rPlots/4C.png', width = 8, height = 3, units = 'in', res = 300)
ComplexHeatmap::draw(hms, row_split = groups, merge_legend = T, heatmap_legend_side = 'right', )
#dev.off()
```

## Fig 4D. 
```{r, 4D}

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("grey60","grey50", 'grey20')

deg.ss.browser <- faire.osaDeGrad.ssPool %>%
  dplyr::bind_rows(., faire.wt.ssPool[faire.wt.ssPool$grp == 'wt.3LW',]) %>%
  dplyr::mutate(color = track.colors)

deg.tracks <- get_cnr_tracks(deg.ss.browser, ylim = c(0,15))
deg.tracks <- c(deg.tracks$wt_3LW, deg.tracks$osaGFP.deGrad_control_faire, deg.tracks$osaGFP.deGrad_nubG4_faire)

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/4D.png', width = 4, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, deg.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```







