# Ideas
- scale data by fraction of effected cells within heterologous tissue?
  If 10% of cells show effect by protein degradation
  - what fold change if any is there at brD? 

# Setup

```{r setup}
library(magrittr)
library(ggplot2)
library(ggrepel)
library(GenomicRanges)

library(org.Dm.eg.db)
library(AnnotationDbi)

library(EnrichedHeatmap)
library(ComplexHeatmap)

library(GO.db)

library(patchwork)

source('utils.R')

###
# load rData
###

load('rData/sheets.rda')
load('rData/peaks.rda')

#TODO move RNAseq data into BAP project
load('/Users/m/McKay/Wing-RNAseq/rOut/wing-rnaseq.RData') 

#TODO move flyfactor database to BAP - code to download to repo?
## load in a flyfacotr fbgn validated .txt - generated with Flybase id validator

## there are 2 'FlyFactor' FBgn codes that are updated two more than 1 unique 'Validated' FBgns in `fbgn.validated`
# FBgn0051782 -- CG31782 in flyfactor - old annotation? no present in Flybase, maybe split into three differen pseudogene annos now
# FBgn0050420 -- is only Atf-2 in flyfactor db
fbgn.validated <- read.table('~/McKay/Motif_databases/FLY/fbgn_validated.txt', header = F, sep = '\t')
colnames(fbgn.validated) <- c('FlyFactor', 'Validated', 'Symbol')
fbgn.validated %<>% 
  dplyr::filter(FlyFactor != 'FBgn0051782') %>% #drop this FBgn, problematic update to multiple new FBgns of pseudogenes
  dplyr::filter(Validated != 'FBgn0265182') # drop this Validated FBgn, which is an incorrectly updated from FBgn0050420
                

###
# assign global variables
###

dm6 <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6
dm3 <- BSgenome.Dmelanogaster.UCSC.dm3::BSgenome.Dmelanogaster.UCSC.dm3
dm6.TxDb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene::TxDb.Dmelanogaster.UCSC.dm6.ensGene

brD <- data.frame('seqnames' = 'chrX',
                  'start' = 1565708,
                  'end' = 1567401) %>%
  GenomicRanges::GRanges()

###
# colors
### 

viridis.hex <- c("#440154","#3b528b","#21918c","#5ec962","#5ec962","#fde725")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")


###
# peak subsets
###

osaPeaks <- peaks %>% 
  dplyr::filter(osa.cnr & !yw.cnr & assay == 'cnr') %>% 
  GRanges() %>% 
  resize(width = 1, fix = "center")


#TODO - this is maybe a bad control because its only 300 sites
osaUnbound <- peaks %>%
  dplyr::filter(yw.cnr & assay == 'cnr') %>% 
  GRanges() %>% 
  resize(width = 1, fix = "center")

#trying this...
background <- peaks %>%
  dplyr::filter(assay == '1kb background') %>%
  GRanges() %>%
  resize(width = 1, fix = 'center')

fairePeaks.wt <- peaks %>%
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(width = 1, fix = "center")

#DONE - remove the peakCat ? since now using log2fold change for these categories...
fairePeaks.deg <- peaks %>%
  dplyr::filter(assay == 'faire' & experiment == 'osaGFP deGrad Pupal Wing FAIRE') %>%
#  dplyr::mutate(deg.peakCat = dplyr::case_when(osaGFP.deGrad & !osaGFP.control ~ 'osa ectopic',
#                                               !osaGFP.deGrad & osaGFP.control ~ 'osa dependent',
#                                               osaGFP.deGrad & osaGFP.control ~ 'osa independent',
#                                               T ~ as.character(NA))) %>%  # NA here would be peaks that aren't reproducible or of good quality
  dplyr::filter(osaGFP.deGrad | osaGFP.control) %>% # want to only use peaks that are reproducible and of good quality 
  GRanges() %>%
  resize(width = 1, fix = "center")


fairePeaks.osaDependent <- fairePeaks.deg %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent') %>%
  GRanges()

fairePeaks.osaIndependent <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Independent') %>%
  GRanges()

fairePeaks.osaEctopic <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Ectopic') %>%
  GRanges()
###
# load bws and matrices for heatmaps
### 

#TODO - cleanup - move matrices?

bws <- get_bws(dplyr::bind_rows(cnr.ss, faire.wt.ssPool), 'id') 

deg.bws <- get_bws(faire.ss %>% dplyr::filter(experiment == 'osaGFP deGrad FAIRE'), 'id')

mats <- purrr::map(bws, ~{
  normalizeToMatrix(., osaPeaks, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
})
names(mats) <- names(bws)

#TODO -- move to utils?
###
# functions 
### 
plot.frac <- function(x) {
 lapply(unique(x$change), function(i) {
   x %>%
     dplyr::filter(change == i) %>%
     ggplot(aes(fraction, change, fill = osa.cnr)) +
     geom_bar(stat = 'identity', width = 1) +
     geom_text(aes(label = per), position = position_fill(vjust = 0.6),  size = 7, fontface = "bold") +
     #geom_text(aes(label = per, position = position_stack(vjust = 0.5))) +
     scale_fill_manual(values = c('grey80','#41f0ad'), name = 'Osa Bound') +
     scale_y_discrete(position = 'right', ) +
     xlab('Percent Bound by Osa') +
     theme(legend.position = 'top',
           legend.key.size = unit(1, 'cm'),
           legend.text = element_text(size = 15),
           legend.title = element_text(size = 20), 
           panel.background = element_blank(),
           axis.text = element_blank(),
           axis.title = element_blank(),
           axis.ticks = element_blank(),
#           axis.text.y = element_text(size = 12),
           panel.grid = element_blank()) + 
     NULL
   fn = paste0('rPlots/3E_',i,'.png')
   ggsave(fn, width = 6, height = 1.5, units = 'in', dpi = 300 )
   })
}



### for gviz
genetrack <- Gviz::GeneRegionTrack(range = dm6.TxDb, fill = 'grey80', showTitle = F, background.title = 'white')
```

TODO - make reference table of different peak categories and parameters, eg. how osa specific peaks are defined, how closing peaks are defined
TODO - currently defining dynamic faire simply by log2fold change - no peak call requirement
# Fig 1

## Fig 1A - brDisc browser shot
```{r, 1A}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("grey20")

faire.wt.browser <- faire.wt.ssPool %>%
  dplyr::filter(grepl('3LW',time)) %>%
  dplyr::mutate(color = track.colors)

faire.wt.tracks <- get_cnr_tracks(faire.wt.browser, ylim = c(0,15))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = '#e858a5', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)


png('rPlots/1A.png', width = 10, height = 2, res = 300, units = 'in')
Gviz::plotTracks(c(axis,faire.wt.tracks, brdisc.at, genetrack), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 20000),
                 to = end(brD + 20000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()

```


# Fig 3


## 3A. heatmap of osa binding sites vs control

```{r, 3A}

#TODO - functionalize normalizetomatrix call, + common min/max and colorRamp

cnr.mats <- list(mats$yw.rep1, mats$osaGFP.rep1, mats$yw.rep2, mats$osaGFP.rep2) # correct order for subsequent names()
names(cnr.mats) <- cnr.ss$id

common_min = min(unlist(cnr.mats))
common_max = max(unlist(cnr.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

osa.count <- osaPeaks %>%
  data.frame() %>%
  nrow()

osa.count[1] <- as.character(osa.count[1])
  
names(osa.count) <- 'Osa Peak'

groups <- ifelse(osaPeaks$osa.cnr, 'Osa Peak', NA)

cnr.hms <- purrr::map(names(cnr.mats), function(x) {
  if(x == 'osaGFP.rep1'){
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,6))),
                    left_annotation = c(rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(osa.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm'))))) 
  } else {
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun, 
                    show_heatmap_legend = FALSE,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis = F, ylim = c(0,6)))) 
  }
})
names(cnr.hms) <- names(cnr.mats)

hms.ordered <-cnr.hms$osaGFP.rep1 + cnr.hms$osaGFP.rep2 + cnr.hms$yw.rep1 + cnr.hms$yw.rep2

png('rPlots/3A.png', width = 6, height = 6, units = 'in', res = 300)
hms.ordered
dev.off()
```

## 3B. Osa CnR brDisc browswer shot

```{r, fig.width=4, fig.height=3}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("#B8E186","#F1B6DA", "#4DAC26","#D01C8B")

cnr.ss.browser <- cnr.ss %>%
  dplyr::mutate(color = track.colors)

cnr.tracks <- get_cnr_tracks(cnr.ss.browser, ylim = c(0,12))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/3B.png', width = 4, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, cnr.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```

## 3C. Where are osaGFP binding sites?
  
```{r, 3C-peak-location}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr | assay == '1kb background') %>%
  dplyr::group_by(assay, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
  ggplot(aes(x = assay, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.3, 'cm'),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/3C.png', width = 4, height = 4, units = 'in', dpi = 300)

```

## 3D. How does osa binding correlate with wt faire data?

```{r 3LW-FAIRE-in-OsaPeaks}


fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks %>%
  data.frame() %>%
  dplyr::group_by(osa.cnr, WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm'))),
                                                 rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'right'))))
  

png('rPlots/3D.png', width = 5, heigh = 5, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, merge_legend = T)
dev.off()


```


## 3E. Fraction of Osa binding sites withing WT 3LW-24h FAIRE categories
```{r, 3E}
#TODO - combine fraction barplots and heatmaps into single figure - cowplots? grid?

faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.wt, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.wt, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat, faire.24h.mat))
common_max = max(c(faire.3lw.mat, faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::count() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')

groups <- fairePeaks.wt$faireCat.3LW_24h

lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                      left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 



png('rPlots/3E.png', width = 4, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hm, annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'left', )
dev.off() 
 
peaks %>% 
  dplyr::filter(assay == 'faire' & !yw.cnr & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::group_by(faireCat.3LW_24h, osa.cnr) %>%
  dplyr::count() %>%
  na.omit() %>%
  reshape2::melt(id = c("osa.cnr","n"), variable.name = 'stage', value.name = 'change') %>%
  dplyr::group_by(stage, change, osa.cnr) %>%
  dplyr::summarise(peaks = sum(n)) %>%
  dplyr::mutate(fraction = peaks / sum(peaks),
                change = factor(change, levels = c('static','opening','closing'))) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()

```

## Drafts
### osa bound closing nearest gene expression dynamics
are genes near osa bound closing sites downregulated?
if any, which?
```{r}
#TODO - use deseq to better identify differential genes -- method below may not be accurate
osaPeaks.closing.downGene <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  data.table::setnames('geneId','Geneid') %>%
  dplyr::select(Geneid, SYMBOL) %>%
#  dplyr::distinct() %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>%
  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
                l3v24 = log2(L3_mean / h24_mean),
                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
                                       l3v24 <= -1 ~ 'up',
                                       T ~ 'static')) 

osaPeaks.closing.downGene %>%
  dplyr::group_by(SYMBOL) %>%
  dplyr::tally() %>%
  dplyr::mutate(symbol = forcats::fct_reorder(SYMBOL, n)) %>%
  dplyr::filter(n>=3) %>%
  ggplot(aes(n, symbol)) +
  geom_point()

osaPeaks.closing.downGene %>%
  dplyr::group_by(SYMBOL) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
closing.go <- select(org.Dm.eg.db, 
       keys = osaPeaks.closing.downGene$Geneid, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

closing.terms <- select(GO.db, 
       keys = closing.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

closing <- dplyr::bind_cols(closing.go, closing.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

closing.entrez <- closing %>% .$ENTREZID

ggo.closing <- clusterProfiler::enrichGO(gene = closing.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.closing %>%
  data.frame() 
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol) %>%
  dplyr::select(symbol) %>%
  dplyr::distinct() %>%
  dplyr::arrange(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
#png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.closing, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
#dev.off()


  #dplyr::group_by(cat) %>%
  #dplyr::tally() %>%
  #ggplot(aes('', n, fill = cat)) + 
  #geom_bar(stat = 'identity', position = 'fill')
#
  #dplyr::group_by(cat,l3v24) %>%
  #dplyr::tally()
  #reshape2::melt() %>%
  #  dplyr::rowwise() 
  #  dplyr::mutate(variable = stringr::str_replace(variable, 'X', ''),
  #                rna_grp = stringr::str_split_fixed(variable, pattern = '_', n = 2)[1],
  #                rna_grp = factor(rna_grp, levels = c('L3','6h','18h','24h','36h','44h'))) %>%
  #dplyr::filter(!is.na(rna_grp)) %>%
  #  dplyr::group_by(cat) %>%
  #dplyr::tally() 
  #ggplot(aes(cat,))
#
  #  dplyr::summarise(mean = mean(value)) %>%
  #  dplyr::mutate(log_mean = log10(1 + mean),
  #                SYMBOL = forcats::fct_reorder(SYMBOL, log_mean)) %>%
  #dplyr::filter(cat == 'down') %>%
  #ggplot(aes(rna_grp, SYMBOL, fill = log_mean)) +
  #geom_raster()
  #geom_point(alpha = 0.5)
  #geom_freqpoly(bins = 10) 
#
  #theme(legend.position = 'none') 
  

```
### motifs in osa peaks
```{r}
osaBound.seq = osaPeaks %>% 
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.streme <- memes::runStreme(osaBound.seq, control = 'shuffle', outdir = 'rPlots/')

purrr::map(osaBound.dreme$motif, function(x) {
  print(x) 
  })

universalmotif::view_motifs)

#universalmotif::view_motifs(osaBound.streme$motif$m7_ATATGTATATATA)

osaBound.streme$motif

osaBound.streme %>%
  dplyr::mutate(consensus = forcats::fct_reorder(consensus, log_evalue)) %>%
  dplyr::slice_min(order_by = log_evalue, n = 5) %>%
  ggplot(aes('', y = consensus, fill = nsites)) + 
  geom_raster()
  geom_point(size = 5)
```
```{r}
osaBound.dreme %>%
  dplyr::mutate(consensus = forcats::fct_reorder(consensus, eval)) %>%
  dplyr::filter(nsites >= 750) %>%
  ggplot(aes(eval, consensus, color = nsites)) + 
  geom_point(size = 5)


osa.tomtom <- memes::runTomTom(osaBound.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 

tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osaBound-dreme_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```

### AT content within osa peaks
Worked on this for a while... not sure exactly what it means, and it seems to contradict other lit. 
But, this gets the sequences around osa peaks out to total of 4kb and then calculates the AT or GC frequency at each base across all peaks.
ie. 50% AT => 50% of peaks of an A or T at that position. 
This plot suggests that at the peak centers ~52-55% of peaks ha

```{r}
osaBound.seq = osaPeaks %>% 
  resize(3000, "center") %>%
  memes::get_sequence(dm6)

osaBound.seq.df <- osaBound.seq %>%
  data.frame() 

colnames(osaBound.seq.df) <- 'seqs'

osaBound.seq.split <- stringr::str_split(osaBound.seq.df$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(osaBound.seq.split) <- as.character(c(1:3000))

osaBound.df <- dplyr::bind_cols(osaBound.seq.split, data.frame(osaPeaks))

osaBound.df %>%
  tidyr::pivot_longer(c(1:3000)) %>%
  dplyr::mutate(name = as.integer(name)) %>%
  #dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
  #              AT = ifelse(value == 'T' | value == 'A', T, F),
  #              name = as.integer(name)) 
  dplyr::group_by(name, value) %>%
  dplyr::tally() %>%
  dplyr::filter(value != 'N') %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  ggplot(aes(name, freq, color = value)) +
  ylim(c(0.15,0.35)) +
  geom_point(alpha = 0.2) +
  scale_color_manual(values = c('#D81B60','#004D40','#FFC107','#1E88E5'))
#  geom_smooth(method = 'loess', span = 0.1) 
ggsave('rPlots/3-draft-ATcontent.png', dpi = 300, width = 5, height = 3)
  #geom_bar(stat = 'identity', position = 'fill')

#####
```


#### promoters control
```{r}
dm6.promoters <- promoters(dm6.TxDb) %>%
  data.frame() %>%
  dplyr::filter(seqnames %in% c('chr2L','chr2R','chr3L','chr3R','chrX')) %>%
  GRanges() %>%
  resize(2000, "center") 

dm6.promoters.seq <- dm6.promoters %>%
  memes::get_sequence(dm6) %>%
  data.frame() 

colnames(dm6.promoters.seq) <- 'seqs'

dm6.promoters.split <- stringr::str_split(dm6.promoters.seq$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(dm6.promoters.split) <- as.character(c(1:2000))

dm6.promoters.df <- dplyr::bind_cols(dm6.promoters.split, data.frame(dm6.promoters))

dm6.promoters.df

dm6.promoters.df %>%
 # dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
  tidyr::pivot_longer(c(1:2000)) %>%
  dplyr::mutate(name = dplyr::case_when(strand == '+' ~ as.integer(name)-1000,
                                        strand == '-' ~ -(as.integer(name)-1000))) %>%
  dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
                AT = ifelse(value == 'T' | value == 'A', T, F),
                name = as.integer(name)) %>%
  dplyr::group_by(name, AT) %>%
  dplyr::tally() %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  dplyr::filter(AT) %>%
  ggplot(aes(name, freq)) +
  ylim(c(0,1)) +
  geom_point(alpha = 0.2) +
  geom_smooth(method = 'loess', span = 0.1)
```

#### at string search
```{r}
#####

osaBound.center.seq = osaPeaks %>% 
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.flank.up.seq = osaPeaks %>%
  shift(250) %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.flank.down.seq = osaPeaks %>%
  shift(-250) %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

# freqIslands = vcountPattern("CG", seqChr8Islands) / width(seqChr8Islands) 
osaBound.center.at <- Biostrings::vcountPattern('AT', osaBound.center.seq) / width(osaBound.center.seq) %>%
  data.frame(values = .)

osaBound.center.at %<>%
  dplyr::mutate(grp = 'center')

osaBound.flank.up.at <- Biostrings::vcountPattern('AT', osaBound.flank.up.seq) / width(osaBound.flank.up.seq) %>%
  data.frame(values = .) 

osaBound.flank.up.at %<>%
  dplyr::mutate(grp = 'flank-up')

osaBound.flank.down.at <- Biostrings::vcountPattern('AT', osaBound.flank.down.seq) / width(osaBound.flank.down.seq) %>%
  data.frame(values = .) 

osaBound.flank.down.at %<>%
  dplyr::mutate(grp = 'flank-down')

test <- dplyr::bind_rows(osaBound.center.at,
                 osaBound.flank.up.at,
                 osaBound.flank.down.at)

test %>%
  ggplot(aes(grp, values)) +
  geom_violin() +
  geom_point(alpha = 0.4, position = position_jitter(width = 0.2))
```

#### other
```{r}


test.mat <- osaBound.seq.split %>%
#osaBound.seq.split %>%
#  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
#  dplyr::filter(!WT.3LW) %>%
  #dplyr::select(id, WT.3LW, c(60:4059)) %>%
#  dplyr::select(c(60:2058)) %>%
  purrr::map(., function(x) {
    dplyr::case_when( x == 'A' | x == 'T' ~ 1, x == 'G' | x == 'C' | x == 'N' ~ 0)
  }) %>%
  dplyr::bind_rows(.) 
  

test.mat %>%
  dplyr::mutate(peak = 1:nrow(.)) %>%
  tidyr::pivot_longer(!peak) %>%
#  dplyr::mutate(value = factor(value)) %>%
  dplyr::group_by(value, name) %>%
  dplyr::tally() %>%
  dplyr::filter(value == 1) %>%
  dplyr::mutate(name = as.integer(name),
                avg = n / 1953) %>%
 # dplyr::arrange(name)
  ggplot(aes(name,avg )) +
  geom_point()
  geom_tile() +
#  scale_fill_manual(values = c('white','red')) +
  theme(axis.text.x = element_blank())

#ComplexHeatmap::Heatmap(test.mat, cluster_rows = F )
#pheatmap::pheatmap(test.mat)
```


#### bedtools nuc control test
writes bed files for regions at center and flanking osa peaks -- need to then run bedtools nuc on those bed files with dm6 fasta
then plots the average AT content within those windows.
Just a sanity check that I can get a similar result to what I have above. 
```{r}
#source('~/NystLib/R/peakUtils.R')
#
##ctbp <- readNarrowPeak('~/McKay/CtBP_CnR/Peaks/yw_3LW_wing_aCtBP_sup-Rep1_dm6_trim_q5_dupsRemoved_allFrags_peaks.narrowPeak') %>%
#
#
#####
osaPeaks.bed <- osaPeaks %>%
  resize(250, 'center') 

osaPeaks %>% shift(500) %>%
  resize(250, 'center') %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_up250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')

osaPeaks %>% shift(-500) %>%
  resize(250, 'center') %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_down250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')
#seqlevelsStyle(osaPeaks.bed) <- 'NCBI'

osaPeaks.bed %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')


center <-  read.csv('osaPeaks_250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'center')

up <-  read.csv('osaPeaks_up250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'up')

down <-  read.csv('osaPeaks_down250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'down')

dplyr::bind_rows(up, center, down) %>%
  dplyr::select(X4_pct_at, cat) %>%
  dplyr::group_by(cat, X4_pct_at) %>%
  ggplot(aes('',X4_pct_at, fill = cat)) +
  geom_violin()
```
#### test control 
```{r}
#k27me3 <- readNarrowPeak('~/McKay/CtBP_CnR/Peaks/yw_3LW_wing_aK27me3_sup-Rep1_dm6_trim_q5_dupsRemoved_allFrags_peaks.narrowPeak') %>%
k27me3 <- readNarrowPeak('~/McKay/ecrGFSTF-3LW-Wing-CnR/Peaks/GSE202807_EcRGFSTF_3LW_WING_CnR_antiFlag_Merge_Sup_dm6_trim_q5_dupsRemoved_20to120_peaks.narrowPeak') %>%
#ctbp <- readNarrowPeak('/Volumes/McKay/Pierre : Lab/snystrom/bioinformatics3001-2_workstation_backup/E93_GOF_Project/ChIP_Data/E93_GOF_ChIP/Peakfiles/E93_GOF_ChIP_2Reps_FAIRE_PooledPeaks_peaks.narrowPeak') %>%
  data.frame() %>%
  dplyr::filter(!seqnames %in% c('chr4','chrY','chrM')) %>%
  GRanges() %>%
  resize(width = 1000, fix = 'center') %>%
  memes::get_sequence(dm6)

k27me3.df <- k27me3 %>%
  data.frame() 

colnames(k27me3.df) <- 'seqs'


k27me3.split <- stringr::str_split(k27me3.df$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(k27me3.split) <- as.character(c(1:1000))

#test <- dplyr::bind_cols(data.frame(osaPeaks), osaBound.seq.split)

k27me3.split %>%
#osaBound.seq.split %>%
  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
  #dplyr::filter(anno.new == "3' UTR") %>%
  #dplyr::select(anno.new, c(59:4058)) %>%
  tidyr::pivot_longer(!id) %>%
  dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
                AT = ifelse(value == 'T' | value == 'A', T, F),
                name = as.integer(name)) %>%
  dplyr::group_by(name, AT ) %>%
  dplyr::tally() %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  dplyr::filter(AT) %>%
  ggplot(aes(name, freq)) +
  geom_point(alpha = 0.4)
#ggsave('rPlots/AT-K27me3_sup.png')
```


```{r}
#
####
#osaUnbound.seq.df <- osaUnbound.seq %>%
#  data.frame() 
#
#colnames(osaUnbound.seq.df) <- 'seqs'
#
#test <- stringr::str_split(osaUnbound.seq.df$seqs, pattern = '', simplify = T) %>%
#  data.table::data.table()
#
#colnames(test) <- as.character(c(1:50))
#
#test %>%
#  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
#  tidyr::pivot_longer(!id) %>%
#  dplyr::group_by(name, value) %>%
#  dplyr::tally() %>%
#  dplyr::mutate(name = as.integer(name),
#                value = factor(value, levels = c('A','T','G','C'))) %>%
#  ggplot(aes(name, n, fill = value)) +
#  geom_bar(stat = 'identity', position = 'fill')
#ggsave('rPlots/TEMP-osaUnbound-ATcontent.png')
#
####
osaBound.seq = osaPeaks %>% 
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.streme.250 <- memes::runStreme(osaBound.seq, control = 'shuffle') 

osa.tomtom <- memes::runTomTom(osa.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 
osa.tomtom
osa.streme$motif
#####
tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osaBound_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```


### motif enrichment in osa dependent regions 
I'm not convinced there is ver meaningful information in motif analysis for osa cnr....
```{r}

###

#memes::
###


osa_closing.seq = fairePeaks.wt %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr & assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

control_closing.seq = fairePeaks.wt %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.3LW_24h == 'closing' & !osa.cnr & assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
 
osa.streme <- memes::runStreme(osa_closing.seq, control = control_closing.seq) 

###
osa.streme %>%
  dplyr::mutate(neg_log_evalue = -as.double(log_evalue),
                consensus = forcats::fct_reorder(consensus, neg_log_evalue, .desc = F)) %>%
  ggplot(aes(x=0,consensus, fill = neg_log_evalue)) +
  geom_tile(color = 'black') +
  scale_fill_viridis_c(guide = guide_legend(direction = 'horizontal', 
                                            nrow = 1, 
                                            label.position = 'bottom', 
                                            title.position = 'top')) +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 12),
        legend.position = 'bottom',
        legend.key.size = unit(0.2,'cm'))
ggsave('rPlots/motif-analysis/3-motif-consensus.png', width = 1.2, height = 2, dpi = 300, scale = 3)
###

#plot of enriched consensus significance
osa.tomtom <- memes::runTomTom(osa.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 


#####
tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osa-bound-closing_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```
### go analysis draft
```{r}
test <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing')
#  dplyr::filter(faire_osaDeGrad.log2FoldChange <= -1) %>%
#  dplyr::arrange(faire_osaDeGrad.log2FoldChange) 
  
 
dependent.go <- select(org.Dm.eg.db, 
       keys = test$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.dependent %>%
  data.frame() %>%
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
dev.off()

```
### osa cnr peak annotation
```{r, 3draft-1.png}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr & faireCat.3LW_24h != 'NA') %>%
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
  ggplot(aes(x = faireCat.3LW_24h, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.01, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1,
                  ) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.3,"cm"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/drafts/3draft-1.png', width = 4, height = 4, dpi = 300)

  
```

### peak annotation split by osa binding and dynamics
```{r, 3draft-2.png}

peaks %>%
  dplyr::filter(assay == 'faire' & faireCat.3LW_24h != 'NA' & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::mutate(osa.specific = ifelse(osa.cnr & !yw.cnr, T, F)) %>%
 # dplyr::filter(grepl('Distal|Exon|Intron|Promoter', anno.new)) %>%
 # dplyr::mutate(anno.temp = dplyr::case_when(grepl('Distal|Intron', anno.new) ~ 'Distal or Intronic',
 #                                           grepl('Exon', anno.new) ~ 'Exon',
 #                                           grepl('Promoter', anno.new) ~ 'Promoter',
 #                                           T~anno.new)) %>%
  dplyr::group_by(osa.specific, anno.new, faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n),
                pct = round(pct, digits = 2)) %>%
  ggplot(aes(x = n, y = anno.new, fill = osa.specific)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text(aes(label = n), position = position_fill(vjust = 0.5), color = 'black') +
  scale_fill_manual(values = c('#D53B59','#52CA3E','#1E82E0')) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        axis.text.x = element_text(size = 5),
        legend.key.size = unit(0.3,"cm")) +
  facet_wrap(~faireCat.3LW_24h)

ggsave('rPlots/drafts/3draft-2.png', width = 6, height = 3, dpi = 300)
```


# Fig 4

There is very little change between control and osa deGrad conditions. 
Figure 4 should demonstrate this and highlight if there are any subtle changes, where they occur, and how they relate to osa binding. 

## 4C - venn

```{r}
osaDeg <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(osaGFP.deGrad) %>%
  GRanges()

control <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(osaGFP.control) %>%
  GRanges()

#ChIPpeakAnno::makeVennDiagram(list(osaDeg,control))
ChIPpeakAnno::makeVennDiagram(faire.osaDeg.byGrp)

####


```

## 4D 
```{r}
### draft log2fold by degrad cat
#Might be useful to articulate that while there are some very subtle differences in osa degrad - there are very few twofold ectopic changes 
#ie. there is not an appreciable gain of accessibility genome-wide when osa is depleted

fairePeaks.deg %>%
  data.frame() %>%
  dplyr::mutate(twofold = ifelse(faire_osaDeGrad.log2FoldChange <= -1 | faire_osaDeGrad.log2FoldChange >= 1, T, F),
                cat = dplyr::case_when(osaGFP.control & !osaGFP.deGrad ~ 'control',
                                       !osaGFP.control & osaGFP.deGrad ~ 'osa degrad',
                                       T ~ 'other')) %>% 
  ggplot(aes(x = '',  faire_osaDeGrad.log2FoldChange, fill = faireCat.osaDeGrad)) + 
  geom_violin(position = 'identity', scale = 'count') +
    geom_point(aes(alpha = twofold), position = position_jitter(width = 0.1), color = 'red') +
  scale_alpha_manual(values = c(0,0.2)) +
  ylim(c(-2.25,2.25)) +
  scale_fill_manual(values = c('#C05066','#8AD04E','#608AE3')) +
  ylab("log2FoldChange(Control / OsaDeGrad)") +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_text(size = 12)) 

ggsave('rPlots/4-draft-5.png', width = 2, height = 4, dpi = 300)

```

## 4D 
```{r}

fairePeaks.deg %>%
  data.frame() %>%
  dplyr::mutate(twofold = ifelse(faire_osaDeGrad.log2FoldChange <= -1 | faire_osaDeGrad.log2FoldChange >= 1, T, F),
                cat = dplyr::case_when(osaGFP.control & !osaGFP.deGrad ~ 'control',
                                       !osaGFP.control & osaGFP.deGrad ~ 'osa degrad',
                                       T ~ 'other')) %>% 
  ggplot(aes(x = '',  faire_osaDeGrad.log2FoldChange, fill = cat)) + 
  geom_violin(position = 'identity', alpha = 0.3, scale = 'count') +
  scale_alpha_manual(values = c(0,0.2)) +
  ylim(c(-2.25,2.25)) +
  scale_fill_manual(values = c('#C05066','#8AD04E','#608AE3')) +
  ylab("log2FoldChange(Control / OsaDeGrad)") +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_text(size = 12)) 

ggsave('rPlots/4-draft-6.png', width = 2, height = 4, dpi = 300)

```


## 4E. FAIRE control vs osaDeGrad - split by degrad categories
Using deseq differential scoring between control and osaGFP deGrad to define FAIRE peak categories.
Using a very sensitive cutoff for log2fold change. 
"ectopic" (gain of signal in degrad) <= log2(control/degrad) <= -0.4, 
"dependent" (loss of signal in degrad) >= log2(control/degrad) >= 0.4,
"independet" (no significant change).


```{r, 4D}

#TODO - are heatmaps misleading? is this difficult to parse when split by these different peak call groups?


# union osa degrad peaks
deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::group_by(faireCat.osaDeGrad) %>% #peak category assigned in peaks.R
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')

groups <- fairePeaks.deg$faireCat.osaDeGrad

lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4B.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'bottom', )
dev.off()

fairePeaks.deg %>% 
  data.frame() %>%
  dplyr::mutate(osa.cnr.specific = ifelse(osa.cnr & !yw.cnr,T,F)) %>%
  dplyr::group_by(faireCat.osaDeGrad, osa.cnr) %>%
  dplyr::tally() 
  dplyr::mutate(fraction = n / sum(n),
                change = faireCat.osaDeGrad) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()


```



## 4F. osa dependent browser shot example
```{r, 4D}
#TODO - add highlights for dependent regions - currently manually annotating in illustrator 

#peaks %>%
#  dplyr::filter(assay == 'faire' & experiment == 'osaGFP deGrad Pupal Wing FAIRE' & faireCat.osaDeGrad == 'Osa Dependent') %>%
#  dplyr::filter(osaGFP.deGrad | osaGFP.control) %>%
#  dplyr::arrange(desc(faire_osaDeGrad.log2FoldChange)) %>%
#  ggplot(aes(faire_osaDeGrad.pvalue, faire_osaDeGrad.log2FoldChange)) +
#  geom_point()
#  dplyr::select(seqnames, start, end) %>%
#  write.table(., file = '~/Desktop/osaDep.bed', col.names = F, row.names = F, quote = F)
  
# chr2R:11,465,947-11,499,493 
inv <- data.frame(seqnames = 'chr2R',
                   start = 11465947,
                   end = 11499493) %>%
  GRanges()
  
#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("#50c42d","#c42d85","grey70","grey60","grey50")

deg.ss.browser <- faire.osaDeGrad.ssPool %>%
  dplyr::bind_rows(., faire.wt.ssPool[faire.wt.ssPool$grp == 'wt.3LW' | faire.wt.ssPool$grp == 'wt.24h' | faire.wt.ssPool$grp == 'wt.44h',]) %>%
  dplyr::mutate(color = track.colors)

deg.tracks <- get_cnr_tracks(deg.ss.browser, ylim = c(0,10))
deg.tracks <- c(deg.tracks$wt_3LW, 
                deg.tracks$wt_24h, 
                deg.tracks$wt_44h,
                deg.tracks$osaGFP.deGrad_control_faire, 
                deg.tracks$osaGFP.deGrad_nubG4_faire)

#brdisc.at <- Gviz::AnnotationTrack(range = inv, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)

png('rPlots/4E-browser.png', width = 4, height = 3, res = 300, units = 'in')
Gviz::plotTracks(c(axis, deg.tracks, genetrack), chromosome = seqnames(inv) %>% as.character(), 
                 from = start(inv),
                 to = end(inv),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```

## 4H. motif enrichment in osa dependent regions 

```{r}

###

#memes::
###


osaDep.seq = fairePeaks.deg %>% 
  data.frame() %>% 
  #dplyr::filter(faireCat.3LW_24h == 'opening' & faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  #dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  #dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
  
osaIndep.seq = fairePeaks.deg %>% 
  data.frame() %>% 
  #dplyr::filter(faireCat.3LW_24h == 'opening' & faireCat.osaDeGrad == 'Osa Independent') %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Independent') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
 
osaDep.streme <- memes::runStreme(osaDep.seq, control = osaIndep.seq) 
#  dplyr::slice_min(pval, n = 8, with_ties = F)

#saDep.streme %<>%
#  dplyr::slice_min(evalue, n = 5, with_ties = F)
osaDep.streme %>%
  data.frame() %>%
  dplyr::mutate(consensus = forcats::fct_reorder(consensus, log_evalue)) %>%
  ggplot(aes(log_evalue, consensus)) + 
  geom_point()


  
###
osaDep.streme %>%
  dplyr::mutate(neg_log_evalue = -as.double(log_evalue),
                consensus = forcats::fct_reorder(consensus, neg_log_evalue, .desc = F)) %>%
  ggplot(aes(x=0,consensus, fill = neg_log_evalue)) +
  geom_tile(color = 'black') +
  scale_fill_viridis_c(guide = guide_legend(direction = 'horizontal', 
                                            nrow = 1, 
                                            label.position = 'bottom', 
                                            title.position = 'top')) +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 12),
        legend.position = 'bottom',
        legend.key.size = unit(0.2,'cm'))
ggsave('rPlots/motif-analysis/4H-consensus.png', width = 1.2, height = 2, dpi = 300, scale = 3)
###

#plot of enriched consensus significance
osaDep.tomtom <- memes::runTomTom(osaDep.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 


#####
tom.results <- tomResults(osaDep.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/',unique(x$consensus),'_osaDependent.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```


## 4I. Osa deGrad FAIRE signal in osa-bound l3-24h closing sites.
This plot shows that within the osa CnR peaks that close from 3LW to 24h, there is little to no change in their ~24h-36h accessibility when osa is depleted.

```{r}
osaClosing <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  GRanges()

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., osaClosing, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

#deg.count <- fairePeaks.deg %>%
#  data.frame() %>%
#  dplyr::group_by(deg.peakCat) %>%
#  dplyr::tally() %>%
#  .$n
#
#names(deg.count) <- c('osa dependent', 'osa ectopic', 'osa independent')

#groups <- osaPeaks$deg.peakCat

#lgd = Legend(at = c("osa dependent", "osa ectopic", "osa independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,1))))
#                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
#                                                                             as.list(deg.count),
#                                                                             side = 'left',
#                                                                             background_gp = gpar(fill = NA, col = NA),
#                                                                             gp = gpar(col = 'black'), 
#                                                                             padding = unit(0, 'mm')),
#                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)))
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(ylim = c(0,1), axis = F)))
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4C.png', width = 3, height = 3, units = 'in', res = 300)
ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```


## 4J. Browser shot of osa deGrad FAIRE at brDisc.
Shows that there is no increase in accessibility.
```{r, 4D}
#TODO - add 36h FAIRE data? better comparison for osa deGrad?

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("grey70","grey60","grey50", "grey40", 'grey30')

deg.ss.browser <- faire.osaDeGrad.ssPool %>%
  dplyr::bind_rows(., faire.wt.ssPool[faire.wt.ssPool$grp == 'wt.3LW' | faire.wt.ssPool$grp == 'wt.24h' | faire.wt.ssPool$grp == 'wt.44h',]) %>%
  dplyr::mutate(color = track.colors)

deg.tracks <- get_cnr_tracks(deg.ss.browser, ylim = c(0,15))
deg.tracks <- c(deg.tracks$wt_3LW, 
                deg.tracks$wt_24h, 
                deg.tracks$wt_44h, 
                deg.tracks$osaGFP.deGrad_control_faire, 
                deg.tracks$osaGFP.deGrad_nubG4_faire)

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/4D.png', width = 4, height = 6, res = 300, units = 'in')
Gviz::plotTracks(c(axis, deg.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```


## Supp. 
Using any peak call from the degrad experiment.
So 'independent' regions are accessible in both experiments. 

```{r}

temp <- fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::filter(faire_osaDeGrad.log2FoldChange >= 0.9) %>%
  #dplyr::filter(faire_osaDeGrad.log2FoldChange >= 0.4 & faire_osaDeGrad.log2FoldChange < 0.9) %>%
  GRanges()
#fairePeaks.osaDependent <- fairePeaks.deg %>% data.frame() %>% dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent')

       #keys = fairePeaks.osaDependent$geneId, 
dependent.go <- select(org.Dm.eg.db, 
      # keys = fairePeaks.osaDependent$geneId, 
       keys = temp$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))


dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
unique(dependent$SYMBOL)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)

#png('rPlots/temp.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
#dev.off()



```

##  4E.

```{r}

osaDependent.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.osaDependent, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(osaDependent.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(osaDependent.mats))
common_max = max(unlist(osaDependent.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

#deg.count <- fairePeaks.deg %>%
#  data.frame() %>%
#  dplyr::group_by(faireCat.osaDeGrad) %>%
#  dplyr::tally() %>%
#  .$n
#
#names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')
#

#TODO - maybe cleaner to have this peak cat assigned globally in setup
groups <- fairePeaks.osaDependent %>% 
  data.frame() %>%
  dplyr::mutate(new.cat <- ifelse(osa.cnr & !yw.cnr, 'Osa Bound', 'Osa Unbound')) %>%
  .$new.cat

lgd = Legend(at = c("Osa Bound", "Osa Unbound"), legend_gp = gpar(fill = 2:3))

deg.hms <- purrr::map(names(osaDependent.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(osaDependent.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(osaDependent.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(osaDependent.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4E.png', width = 4, height = 3, units = 'in', res = 300)
#ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```
## 4F.

```{r}

osaEctopic.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.osaEctopic, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(osaEctopic.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(osaEctopic.mats))
common_max = max(unlist(osaEctopic.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- fairePeaks.osaEctopic %>% 
  data.frame() %>%
  dplyr::mutate(new.cat <- ifelse(osa.cnr & !yw.cnr, 'Osa Bound', 'Osa Unbound')) %>%
  .$new.cat

lgd = Legend(at = c("Osa Bound", "Osa Unbound"), legend_gp = gpar(fill = 2:3))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(osaEctopic.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4F.png', width = 4, height = 3, units = 'in', res = 300)
#ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```



## 4H.

```{r}
#TODO - simplifiy - purrr::map
faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')
#
groups <- fairePeaks.osaDependent$faireCat.3LW_24h

#
lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                     left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.6h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '6h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) + 
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 

png('rPlots/4H.png', width = 4, height = 8, units = 'in', res = 300)
ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
dev.off() 
```

## 4I.
```{r}

faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.osaEctopic %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')
#
groups <- fairePeaks.osaEctopic$faireCat.3LW_24h

#
lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                     left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.6h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '6h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) + 
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 

png('rPlots/4I.png', width = 4, height = 8, units = 'in', res = 300)
ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
dev.off() 
```

```{r}
fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::filter(!anno.new == "3' UTR") %>% #filtering this out because there are no peaks anno as 3'UTR in ectopic regins, messes up colors
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x= n, y = '', fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.background = element_blank(),) +
        #strip.text = element_blank())+
  facet_wrap(~faireCat.3LW_24h, ncol = 1)
ggsave('rPlots/4-draft-1.png', width = 4, height = 5, units = 'in', dpi = 300)
```
```{r}
fairePeaks.osaEctopic %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x = n, y = '', fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.background = element_blank(),) +
        #strip.text = element_blank())+
  facet_wrap(~faireCat.3LW_24h, ncol = 1)
ggsave('rPlots/4-draft-2.png', width = 4, height = 5, units = 'in', dpi = 300)
```


## drafts

### peak annotations
```{r}
fairePeaks.deg %>%
    data.frame() %>%
#  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' ) %>%
  dplyr::group_by(faireCat.osaDeGrad, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x = faireCat.osaDeGrad, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =4,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/4G.png', width = 6, height = 5, units = 'in', dpi = 300)
```
```{r}
#fairePeaks.osaDep <- fairePeaks.deg %>% 
#  data.frame() %>%
#  dplyr::filter(faireCat.3LW_24h == 'opening') %>%
#  GRanges()
#
#faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#
#common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
#common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
#
#col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)
#
##faire.count <- fairePeaks.osaDep %>%
##  data.frame() %>%
##  dplyr::group_by(faireCat.3LW_24h) %>%
##  dplyr::count() %>%
##  dplyr::mutate(n = as.character(n)) %>%
##  .$n
##
##names(faire.count) <- c('closing','opening','static')
##
#groups <- fairePeaks.osaDep$faireCat.osaDeGrad
#
##
#lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))
#                             
#hm <- EnrichedHeatmap(faire.3lw.mat, 
#                      pos_line = F, 
#                      axis_name = c(-2, 0, 2),
#                      col = col_fun,
#                      column_title = '3LW',
#                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
#                                                                               ylim = c(0,6),
#                                                                               axis_param = list(side = 'left'))),  
#                      name = 'FAIRE zScore',
#                     left_annotation = rowAnnotation(#textbox = anno_textbox(align_to = groups,
#                                                     #                        as.list(faire.count),
#                                                     #                        side = 'left',
#                                                     #                        background_gp = gpar(fill = NA, col = NA),
#                                                     #                        gp = gpar(col = 'black'), 
#                                                     #                        padding = unit(0, 'mm')),
#                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
#                      row_title  = NULL,
#                      use_raster = F) +
# EnrichedHeatmap(faire.6h.mat, 
#                 pos_line = F, 
#                 axis_name = c(-2, 0, 2),
#                 col = col_fun,
#                 column_title = '6h',
#                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
#                 row_title  = NULL,
#                 show_heatmap_legend = F,
#                 use_raster = F) + 
# EnrichedHeatmap(faire.24h.mat, 
#                 pos_line = F, 
#                 axis_name = c(-2, 0, 2),
#                 col = col_fun,
#                 column_title = '24h',
#                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
#                 row_title  = NULL,
#                 show_heatmap_legend = F,
#                 use_raster = F) 
# 
#
##png('rPlots/4I.png', width = 4, height = 8, units = 'in', res = 300)
#ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
##dev.off() 
```
### browser draft




### go analysis draft
```{r}
test <- fairePeaks.wt %>%
  data.frame() %>%
#  dplyr::filter(faire_osaDeGrad.log2FoldChange <= -1) %>%
  dplyr::arrange(faire_osaDeGrad.log2FoldChange) 
  


 
fairePeaks.osaEctopic %>% data.frame
dependent.go <- select(org.Dm.eg.db, 
       keys = test$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.dependent %>%
  data.frame() %>%
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(SYMBOL~., scale = 'free')
dev.off()

```




