```{r}
library(magrittr)
library(GenomicRanges)
library(ggplot2)
load('rData/sheets.rda')
source('~/NystLib/R/peakUtils.R')
source('utils.R')


cnr.ss


#faire.ss <- dplyr::bind_rows(faire.osaDeGrad.ss, faire.wt.ss)
faire.peaks <- getPeakData(faire.ss, by = 'grp', narrowPeak_colname = 'peaks')


faire.ss
mcols(faire.peaks %>% GRanges())


faire.peaks <- getPeakData(faire.ss, by = 'id', narrowPeak_colname = 'peaks')
faire.byGrp <- faire.peaks %>% GRanges() %>% split(., mcols(.)$grp)
faire.byEx <- faire.peaks %>% GRanges() %>% split(., mcols(.)$experiment)

faire.union <- faire.byGrp %>%
  unlist() %>%
  reduce()

faire.byGrp <- lapply(faire.byGrp, function(x) grp_qFilter(x, quantile = 0.75, operation = 'subsetByOverlaps'))


### get CUT&RUN peaks
cnr.peaks <- getPeakData(cnr.ss, by = 'id', narrowPeak_colname = 'peak_allFrags')
cnr.byID <- cnr.peaks %>% GRanges() %>% split(., mcols(.)$id)
cnr.byGrp <- cnr.peaks %>% GRanges() %>% split(., mcols(.)$grp)


cnr.byGrp <- lapply(cnr.byGrp, function(x) grp_qFilter(x, quantile = 0.75, operation = 'subsetByOverlaps'))

cnr.union <- cnr.byID %>%
  unlist() %>%
  reduce()
```





```{r}
f.df <- faire.union %>% data.frame() %>% dplyr::mutate(assay = 'faire')
c.df <- cnr.union %>% data.frame() %>% dplyr::mutate(assay = 'cnr')

peaks <- dplyr::bind_rows(f.df, c.df) %>%
  dplyr::filter(!seqnames %in% c('chr4','chrY','chrM')) %>% #drop regions from potentially problematic chromosomes
  dplyr::arrange(seqnames, start, end) %>%
  dplyr::mutate(WT.3LW = ifelse(GRanges(.) %over% faire.byGrp$wt.3LW, T, F),
                WT.6h = ifelse(GRanges(.) %over% faire.byGrp$wt.6h, T, F),
                WT.24h = ifelse(GRanges(.) %over% faire.byGrp$wt.24h, T, F),
                WT.44h = ifelse(GRanges(.) %over% faire.byGrp$wt.44h, T, F),
                osaGFP.control = ifelse(GRanges(.) %over% faire.byGrp$osaGFP.deGrad_control_faire, T, F),
                osaGFP.deGrad = ifelse(GRanges(.) %over% faire.byGrp$osaGFP.deGrad_nubG4_faire, T, F),
                osa.cnr = ifelse(GRanges(.) %over% cnr.byGrp$osaGFP.sup, T, F),
                yw.cnr = ifelse(GRanges(.) %over% cnr.byGrp$yw.sup, T, F),
                cnr.peakCat = dplyr::case_when(osa.cnr & yw.cnr ~ 'shared',
                                               osa.cnr & !yw.cnr ~ 'osa',
                                               !osa.cnr & yw.cnr ~ 'control',
                                               T ~ 'other')) %>%
  dplyr::filter(assay == 'faire' & WT.3LW |
                  assay == 'faire' & WT.6h | 
                  assay == 'faire' & WT.24h | 
                  assay == 'faire' & WT.24h | 
                  assay == 'faire' & WT.44h | 
                  assay == 'faire' & osaGFP.control | 
                  assay == 'faire' & osaGFP.deGrad | 
                  assay == 'cnr' & osa.cnr | 
                  assay == 'cnr' & yw.cnr) %>% #drop regions that don't have a good quality reproducible peak called
#get osaGFP deGrad cnr bigwig scores...
  dplyr::mutate(peak = 1:nrow(.),
                osaGFP.rep1.spikeNorm = get_bw_score(cnr.ss[cnr.ss$baseName == 'osaGFP-3LW-wing-aGFP-CnR-sup-rep1',]$bigwig_allFrags_sacCer3_spikeNorm[1], GRanges(.)),
                osaGFP.rep2.spikeNorm = get_bw_score(cnr.ss[cnr.ss$baseName == 'osaGFP-3LW-wing-aGFP-CnR-sup-rep2',]$bigwig_allFrags_sacCer3_spikeNorm[1], GRanges(.)),
                yw.rep1.spikeNorm = get_bw_score(cnr.ss[cnr.ss$baseName == 'yw-3LW-wing-aGFP-CnR-sup-rep1',]$bigwig_allFrags_sacCer3_spikeNorm[1], GRanges(.)),
                yw.rep2.spikeNorm = get_bw_score(cnr.ss[cnr.ss$baseName == 'yw-3LW-wing-aGFP-CnR-sup-rep2',]$bigwig_allFrags_sacCer3_spikeNorm[1], GRanges(.)),
#get faire bigwig scores... 
                faire.3LW.zScore = get_bw_score(faire.wt.ssPool[faire.wt.ssPool$baseName == 'wt_3LW',]$bigwig_rpgcNorm_zNorm[1], GRanges(.)),
                faire.6h.zScore = get_bw_score(faire.wt.ssPool[faire.wt.ssPool$baseName == 'wt_6h',]$bigwig_rpgcNorm_zNorm[1], GRanges(.)),
                faire.24h.zScore = get_bw_score(faire.wt.ssPool[faire.wt.ssPool$baseName == 'wt_24h',]$bigwig_rpgcNorm_zNorm[1], GRanges(.)),
                faire.44h.zScore = get_bw_score(faire.wt.ssPool[faire.wt.ssPool$baseName == 'wt_44h',]$bigwig_rpgcNorm_zNorm[1], GRanges(.)))
```


```{r}
#DONE - download BAM files
#TODO - confirm all deseq parameters 
cnr.Counts <- Rsubread::featureCounts(cnr.ss$bam,
                        annot.ext = makeSAF(peaks), #trying all peaks - FAIRE and CnR - so it will be easy to combine with main peaks df
                        allowMultiOverlap = T,
                        nthreads = 4,
                        isPairedEnd = T)

colnames(cnr.Counts$counts) <- cnr.ss$id

cnr.dds <- DESeq2::DESeqDataSetFromMatrix(countData = cnr.Counts$counts,  colData = cnr.ss,  design = ~grp) %>% DESeq2::DESeq(.)

cnr.dds.df <- DESeq2::results(cnr.dds, contrast = c('grp','osaGFP.sup','yw.sup')) %>% data.frame() 

names(cnr.dds.df) <- paste0('cnr.',names(cnr.dds.df))

##### 
faire.ss.WT <- faire.ss[faire.ss$experiment == 'WT FAIRE Wing Timecourse',]

faire.Counts <- Rsubread::featureCounts(faire.ss.WT$bam, #just use Bam from WT timecourse - single-end data
                        annot.ext = makeSAF(peaks), #trying all peaks - FAIRE and CnR - so it will be easy to combine with main peaks df
                        allowMultiOverlap = T,
                        nthreads = 4,
                        isPairedEnd = F)

colnames(faire.Counts$counts) <- faire.ss.WT$id

faire.dds <- DESeq2::DESeqDataSetFromMatrix(countData = faire.Counts$counts,  colData = faire.ss.WT,  design = ~grp) %>%
  DESeq2::DESeq(.)

faire.dds.df <- DESeq2::results(faire.dds, contrast = c('grp','wt.3LW','wt.24h')) %>% data.frame() 

names(faire.dds.df) <- paste0('faire_3LW_24h.',names(faire.dds.df))

##### bind deseq results to main peaks dataframe and annotate faire behavior between 3LW and 24hAPF 
# closing --> must be from FAIRE data, must be called a reproducible peak at 3LW, must have a log2(3LW/24h) >= 1 
# static --> must be from FAIRE data, no peak call requirement, must have -1 > log2(3LW/24h) < 1
# opening --> must be from FAIRE data, must be a reproducible peak call at 24h, must have a log2(3LW/24h) <= -1

peaks <- dplyr::bind_cols(peaks, cnr.dds.df, faire.dds.df) %>%
  dplyr::mutate(faireCat.3LW_24h = dplyr::case_when(assay == 'faire' & WT.3LW &  faire_3LW_24h.log2FoldChange >= 1 ~ 'closing',
                                                    assay == 'faire' & faire_3LW_24h.log2FoldChange < 1 & faire_3LW_24h.log2FoldChange > -1 ~ 'static',
                                                    assay == 'faire' & WT.24h & faire_3LW_24h.log2FoldChange <= -1 ~ 'opening',
                                                    T ~ 'NA'))
peaks
peaks %>% dplyr::filter(faireCat.3LW_24h == 'static' & assay == 'faire' )
```



