---
editor_options: 
  markdown: 
    wrap: 72
---

# Ideas

-   scale data by fraction of effected cells within heterologous tissue?
    If 10% of cells show effect by protein degradation
    -   what fold change if any is there at brD?

# Setup

```{r setup}
library(magrittr)
library(ggplot2)
library(ggrepel)
library(GenomicRanges)

library(org.Dm.eg.db)
library(AnnotationDbi)

library(EnrichedHeatmap)
library(ComplexHeatmap)

library(GO.db)

library(patchwork)

source('utils.R')
source('~/NystLib/R/peakUtils.R') #TODO move/rewrite getPeakData from NystLib to utils.R
###
# load rData
###

load('rData/sheets.rda')
load('rData/peaks.rda')

#TODO move RNAseq data into BAP project
# TODO this is in progress - mixed across directories
load('rData/wing-rnaseq-GeTMM.RData')  # GeTMM normalized RNAseq
#loads 'wing.rnaseq'

#load('rData/wing-rnaseq-DDS.RData')  # loads dds.results 
## TODO -- no id validation yet need to fix
#rna.dds <- dds.results %>%
#  dplyr::filter(baseMean_6h > 0) %>%
#  tibble::rownames_to_column(var = 'Geneid')

ON.genes <- wing.rnaseq %>%
  dplyr::filter(mean >= 2) %>%
  dplyr::select(Geneid) 

ON.genes %>% write.table('on-rnaseq.txt', sep = '\n', row.names = F, col.names = F, quote = F)
#TODO clean this up - had to write out the on list and use the flybase id validator tool... 
ON.genes.validated <- read.table('~/McKay/Wing-RNAseq/on-genes-IDvalidation.txt', header = T) %>%
  dplyr::rename(Geneid = submitted_item, 
                Validated = validated_id,
                Symbol = current_symbol)

#TODO move flyfactor database to BAP - code to download to repo?
## load in a flyfacotr fbgn validated .txt - generated with Flybase id validator

## there are 2 'FlyFactor' FBgn codes that are updated two more than 1 unique 'Validated' FBgns in `fbgn.validated`
# FBgn0051782 -- CG31782 in flyfactor - old annotation? no present in Flybase, maybe split into three differen pseudogene annos now
# FBgn0050420 -- is only Atf-2 in flyfactor db
fbgn.validated <- read.table('~/McKay/Motif_databases/FLY/fbgn_validated.txt', header = F, sep = '\t')
colnames(fbgn.validated) <- c('FlyFactor', 'Validated', 'Symbol')
fbgn.validated %<>% 
  dplyr::filter(FlyFactor != 'FBgn0051782') %>% #drop this FBgn, problematic update to multiple new FBgns of pseudogenes
  dplyr::filter(Validated != 'FBgn0265182') # drop this Validated FBgn, which is an incorrectly updated from FBgn0050420
                

###
# assign global variables
###

dm6 <- BSgenome.Dmelanogaster.UCSC.dm6::BSgenome.Dmelanogaster.UCSC.dm6
dm3 <- BSgenome.Dmelanogaster.UCSC.dm3::BSgenome.Dmelanogaster.UCSC.dm3
dm6.TxDb <- TxDb.Dmelanogaster.UCSC.dm6.ensGene::TxDb.Dmelanogaster.UCSC.dm6.ensGene

dm6.promoters <- ChIPseeker::getPromoters(dm6.TxDb, upstream = 500, downstream = 100) 
dm6.genes <- GenomicFeatures::genes(dm6.TxDb)

brD <- data.frame('seqnames' = 'chrX',
                  'start' = 1565708,
                  'end' = 1567401) %>%
  GenomicRanges::GRanges()

shn <- data.frame('seqnames' = 'chr2R',
                  'start' = 11159086,
                  'end' = 11217394) %>% 
  GRanges()

sbb <- data.frame('seqnames' = 'chr2R',
                  'start' = 18278198,
                  'end' = 18357296) %>%
  GRanges()

# Dl 
#chr3R:19,265,389-19,350,508
Dl <- data.frame('seqnames' = 'chr3R',
                  'start' = 19265389,
                  'end' = 19350508) %>%
  GRanges()


# E(spl)-c
#chr3R:25,980,255-26,065,374
espl <- data.frame('seqnames' = 'chr3R',
                  'start' = 25980255,
                  'end' = 26065374) %>%
  GRanges()

### 
# meme db 
###
#TODO - so clunky...
# loading in the fly factor database, drop out a couple of problematic fbgn
# then use the fbgn.validated lookup table for fly factor to update all the old ids to correct ids
# then filter on the on gene list -- which has to go through a similar process of id validation
# gross
meme_db <- universalmotif::read_meme('~/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  universalmotif::to_df() %>%
  dplyr::group_by(name) %>%
  dplyr::filter(!grepl('FBgn0051782|FBgn0265182', name)) %>%  #drop this FBgn, problematic update to multiple new FBgns of pseudogenes
  dplyr::mutate(basename = stringr::str_split_fixed(name, '_', n = 2)[[1]],
                fbgn.validated = fbgn.validated[fbgn.validated$FlyFactor == basename,]$Validated) 

meme_db_on <- meme_db %>%
  dplyr::filter(fbgn.validated %in% ON.genes.validated$Validated)

options(meme_db = universalmotif::to_list(meme_db_on, extrainfo = F))
###
# colors
### 

viridis.hex <- c("#440154","#3b528b","#21918c","#5ec962","#5ec962","#fde725")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
cbPalette.4 <- c("#D81B60","#1E88E5","#FFC107","#004D40")
cbPalette.ibm <- c('#332288','#117733','#44AA99','#88CCEE','#DDCC77','#CC6677','#AA4499','#882255')
###
# peak subsets
###

osaPeaks <- peaks$cnr.df %>% 
  dplyr::filter(osa.cnr & !yw.cnr) %>% 
  GRanges() %>% 
  resize(width = 1, fix = "center")


# REMOVED- this is maybe a bad control because its only 300 sites
#osaUnbound <- peaks %>%
#  dplyr::filter(yw.cnr & assay == 'cnr') %>% 
#  GRanges() %>% 
#  resize(width = 1, fix = "center")

#trying this...
#background <- peaks %>%
#  dplyr::filter(assay == '1kb background') %>%
#  GRanges() %>%
#  resize(width = 1, fix = 'center')

#closing.osaBound <- peaks %>%
#  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse' & faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr) %>%
#  GRanges() %>%
#  resize(width = 1, fix = "center")

fairePeaks.wt <- peaks$faire.wt.df %>%
#  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(width = 1, fix = "center")

#redundant... but used in a couple chunks for motif analysis
fairePeaks <- peaks$faire.wt.df %>% 
  dplyr::filter(assay == 'faire'  & experiment == 'WT FAIRE Wing Timecourse')


#DONE - remove the peakCat ? since now using log2fold change for these categories...
fairePeaks.deg <- peaks$faire.osaDeg.df %>%
#  dplyr::filter(assay == 'faire' & experiment == 'osaGFP deGrad Pupal Wing FAIRE') %>%
  dplyr::filter(osaGFP.deGrad | osaGFP.control) %>% # want to only use peaks that are reproducible and of good quality 
  GRanges() %>%
  resize(width = 1, fix = "center")

fairePeaks.osaDependent <- fairePeaks.deg %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent') %>%
  GRanges()

fairePeaks.osaIndependent <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Independent') %>%
  GRanges()

fairePeaks.osaEctopic <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Ectopic') %>%
  GRanges()
###
# load bws and matrices for heatmaps
### 

#TODO - cleanup - move matrices?

bws <- get_bws(dplyr::bind_rows(cnr.ss, faire.wt.ssPool), 'id') 

deg.bws <- get_bws(faire.ss %>% dplyr::filter(experiment == 'osaGFP deGrad FAIRE'), 'id')

#mats <- purrr::map(bws, ~{
#  normalizeToMatrix(., osaPeaks, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
#})
#names(mats) <- names(bws)

#TODO -- move to utils?
###
# functions 
### 
plot.frac <- function(x) {
 lapply(unique(x$change), function(i) {
   x %>%
     dplyr::filter(change == i) %>%
     ggplot(aes(fraction, change, fill = Osa.Bound)) +
     geom_bar(stat = 'identity', width = 1) +
     geom_text(aes(label = per), position = position_fill(vjust = 0.6),  size = 7, fontface = "bold") +
     #geom_text(aes(label = per, position = position_stack(vjust = 0.5))) +
     scale_fill_manual(values = c('grey80','#41f0ad'), name = 'Osa Bound') +
     scale_y_discrete(position = 'right', ) +
     xlab('Percent Bound by Osa') +
     theme(legend.position = 'top',
           legend.key.size = unit(1, 'cm'),
           legend.text = element_text(size = 15),
           legend.title = element_text(size = 20), 
           panel.background = element_blank(),
           axis.text = element_blank(),
           axis.title = element_blank(),
           axis.ticks = element_blank(),
#           axis.text.y = element_text(size = 12),
           panel.grid = element_blank()) + 
     NULL
   fn = paste0('rPlots/3E_',i,'.png')
   ggsave(fn, width = 6, height = 1.5, units = 'in', dpi = 300 )
   })
}

osaPeaks %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table('osaPeaks.bed', quote = F, col.names = F, row.names = F, sep = '\t')

peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(yw.cnr) %>%
  dplyr::select(seqnames, start, end) %>%
  write.table('controlPeaks.bed', quote = F, col.names = F, row.names = F, sep = '\t')

fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table('osaDependent.bed', quote = F, col.names = F, row.names = F, sep = '\t')

fairePeaks.osaEctopic %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table('osaEctopic.bed', quote = F, col.names = F, row.names = F, sep = '\t')

# write out individual replicate bed files for cnr
# TODO make a bed directory
purrr::map(names(cnr.byID), function(x) {
  path = paste0('rPlots/',x,'.bed')
  
  cnr.byID[x] %>%
    data.frame() %>%
    dplyr::select(seqnames, start, end) %>%
    write.table(path, col.names = F, row.names = F, quote = F, sep = '\t')
})

purrr::map(names(cnr.byGrp), function(x) {
  path = paste0('rPlots/',x,'.bed')
  cnr.byGrp[[x]] %>%
    data.frame() %>%
    dplyr::select(seqnames, start, end) %>%
    write.table(path, col.names = F, row.names = F, quote = F, sep = '\t')
})
### for gviz
genetrack <- Gviz::GeneRegionTrack(range = dm6.TxDb, fill = 'grey80', showTitle = F, background.title = 'white', stacking = 'squish')


```

TODO - make reference table of different peak categories and parameters,
eg. how osa specific peaks are defined, how closing peaks are defined
TODO - currently defining dynamic faire simply by log2fold change - no
peak call requirement \# Fig 1

## Fig 1A - brDisc browser shot

```{r, 1A}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("grey20")

faire.wt.browser <- faire.wt.ssPool %>%
  dplyr::filter(grepl('3LW',time)) %>%
  dplyr::mutate(color = track.colors)

faire.wt.tracks <- get_cnr_tracks(faire.wt.browser, ylim = c(0,15))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = '#e858a5', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)


png('rPlots/1A.png', width = 10, height = 2, res = 300, units = 'in')
Gviz::plotTracks(c(axis,faire.wt.tracks, brdisc.at, genetrack), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 20000),
                 to = end(brD + 20000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()

```
### 1A - draft
```{r}
track.colors <- c("grey20", "grey60")

faire.wt.browser <- faire.wt.ssPool %>%
  dplyr::filter(grepl('3LW|44h',time)) %>%
  dplyr::mutate(color = track.colors) 

faire.wt.tracks <- get_cnr_tracks(faire.wt.browser, ylim = c(0,15))
faire.wt.tracks <- c(faire.wt.tracks$wt_3LW,  faire.wt.tracks$wt_44h)

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = '#e858a5', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)


png('rPlots/1A-allTime.png', width = 4, height = 2, res = 300, units = 'in')
Gviz::plotTracks(c(axis,faire.wt.tracks, brdisc.at, genetrack), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 10000),
                 to = end(brD + 10000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()

```

## 1C. Gene expression changes of mur2b and br

```{r}
wing.rnaseq %>%
  dplyr::select(-log_mean,-mean) %>%
  tidyr::pivot_longer(!Geneid) %>%
  dplyr::filter(Geneid %in% c('FBgn0025390','FBgn0283451')) %>%
  dplyr::group_by(name) %>%
  dplyr::mutate(name = stringr::str_remove(name,'X'),
                grp = stringr::str_split_fixed(name,'_',n=2)[[1]],
                grp = factor(grp, levels = c('L3', '6h','18h','24h','36h','44h')),
                Gene = dplyr::case_when(Geneid == 'FBgn0025390' ~ 'mur2b',
                                        Geneid == 'FBgn0283451' ~ 'br')) %>%
  dplyr::group_by(Gene, grp) %>%
  dplyr::summarise(mean = mean(value)) %>%
  dplyr::mutate(fracmax = mean/max(mean)) %>%
  ggplot(aes(grp, fracmax, color = Gene, group = Gene)) +
  geom_line(size = 2) +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 20),
        panel.grid.major = element_line(color = 'grey80'),
        legend.key.size = unit(1, 'cm'),
        legend.title = element_text(size = 20),
        legend.text = element_text(size = 20))

ggsave('rPlots/1C.png', dpi = 300, width = 7, height = 5)
```

# Fig 3

## 3B. Venn of osa peaks vs control

Manually draw venn in illustrator

```{r}

ChIPpeakAnno::makeVennDiagram(list(peaks$cnr.df %>% dplyr::filter(yw.cnr) %>% GRanges(),
                                   peaks$cnr.df %>% dplyr::filter(osa.cnr) %>% GRanges()))

###

peaks$cnr.df %>%
  dplyr::filter(yw.cnr | osa.cnr) %>%
  dplyr::mutate(twofold = ifelse(log2FoldChange <= -1 | log2FoldChange >= 1, T, F),
                cat = dplyr::case_when(osa.cnr & !yw.cnr ~ 'Osa',
                                       !osa.cnr & yw.cnr ~ 'Control',
                                       T ~ 'Shared')) %>% 
  ggplot(aes(x = '',  log2FoldChange, fill = cat)) + 
#  ggplot(aes(x = '',  cnr.log2FoldChange)) + 
  geom_violin(position = 'identity', alpha = 0.3, scale = 'count') +
  scale_alpha_manual(values = c(0,0.2)) +
#  ylim(c(-2.25,2.25)) +
  scale_fill_manual(values = c('#C05066','#8AD04E','#608AE3')) +
  ylab("log2FoldChange(osaGFP / control)") +
  theme(#legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_text(size = 12)) 

ggsave('rPlots/3B-draft.png', width = 5, height = 5, dpi = 300)

peaks$cnr.df %>%
  dplyr::filter(!yw.cnr & osa.cnr) %>%
  dplyr::filter(log2FoldChange < 0)
```
### test of log2fold filtering
```{r}
peaks$cnr.df%>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  dplyr::filter(log2FoldChange >= 0) %>%
  dplyr::select(seqnames, start, end) %>%
  write.table('logCutOff-osaPeaks.bed', col.names = F, row.names = F, quote = F, sep = '\t')

peaks$cnr.df%>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  #dplyr::filter(log2FoldChange >= 0) %>%
  dplyr::select(seqnames, start, end, log2FoldChange, padj) %>%
  write.table('osaPeaks.bed', col.names = F, row.names = F, quote = F, sep = '\t')
```
### what about peak width between control and sample
At first it looked like there was a skew to narrow peaks in control vs osaGFP, but I realized that 
was due to how peaks were overlapped when finding reproducible peaks - only coordinates of the query were kept, not query and subject.
yw.rep1 has a large skew to narrow peaks, but not rep2. 
A union of the reproducible replicate coordinates has a peak distribution that is very similar to osaGFP. 
ie. its not necessarily true that control peaks are frequently a more narrow sharp signal relative to broader osaGFP signal.
It may still be true that "non reproducible" control peaks have a different shape...
```{r}
 # .byGrp are filtered and replicate reproduced, and reduced peak lists 

dplyr::bind_rows(cnr.byGrp$osaGFP.sup %>% data.frame() %>% dplyr::mutate(tempid = 'osa'),
                 cnr.byGrp$yw.sup %>% data.frame() %>% dplyr::mutate(tempid = 'yw')) %>% 
  dplyr::select(tempid, width) %>%
  dplyr::mutate(tempid = factor(tempid, levels = c('osa','yw'))) %>%
  ggplot(aes(width, color = tempid)) +
  geom_density(aes(y = after_stat(scaled)), size = 2) +
  #geom_freqpoly(aes(y = stat(ncount)), position = 'identity', binwidth = 25, size = 1) +
#  scale_fill_manual(values = c('#fa4bc3', 'grey40'))+
  xlim(c(0,2000)) + 
  xlab('Peak Width (bp)') + 
  ylab('Fraction of Max Bin Count') +
  theme(axis.text = element_text(size = 12), 
        axis.title = element_text(size = 12),
        legend.title = element_blank())
ggsave('rPlots/3-peak_width_histogram.png', width = 4, height = 3, dpi = 300)


## distribution of peak width by peak cat
dplyr::bind_rows(cnr.byID$osaGFP.rep1 %>% data.frame() %>% dplyr::mutate(tempid = 'osa.1'),
                 cnr.byID$osaGFP.rep2 %>% data.frame() %>% dplyr::mutate(tempid = 'osa.2'),
                 cnr.byID$yw.rep1 %>% data.frame() %>% dplyr::mutate(tempid = 'yw.1'),
                 cnr.byID$yw.rep2 %>% data.frame() %>% dplyr::mutate(tempid = 'yw.2')) 

dplyr::bind_rows(cnr.byGrp$osaGFP.sup %>% data.frame() %>% dplyr::mutate(tempid = 'osa'),
                 cnr.byGrp$yw.sup %>% data.frame() %>% dplyr::mutate(tempid = 'yw')) %>% 
  dplyr::mutate(cnr.peakCat = dplyr::case_when(GRanges(.) %over% GRanges(peaks$cnr.df %>% dplyr::filter(!yw.cnr & osa.cnr)) ~ 'osa',
                                               GRanges(.) %over% GRanges(peaks$cnr.df %>% dplyr::filter(yw.cnr & !osa.cnr)) ~ 'control', 
                                               GRanges(.) %over% GRanges(peaks$cnr.df %>% dplyr::filter(yw.cnr & osa.cnr)) ~ 'shared')) %>% 
  na.omit() %>%
  ggplot(aes(width, color = tempid)) +
  geom_density(aes(y = after_stat(scaled)), size = 2) +
#  geom_freqpoly(aes(y = stat(ncount)), position = 'identity', binwidth = 50, alpha = 0.5, size = 2) +
#  scale_fill_viridis_b()+
  xlim(c(0,1500)) + 
  xlab('Peak Width (bp)') + 
 # ylab('Fraction of Max Bin Count') +
  theme(axis.text = element_text(size = 12), 
        axis.title = element_text(size = 12),
        legend.title = element_blank()) +
  facet_wrap(~cnr.peakCat)

## average width calc
dplyr::bind_rows(cnr.byGrp$osaGFP.sup %>% data.frame() %>% dplyr::mutate(tempid = 'osa'),
                 cnr.byGrp$yw.sup %>% data.frame() %>% dplyr::mutate(tempid = 'yw')) %>%
  dplyr::select(tempid, width)  %>%
  dplyr::group_by(tempid) %>%
  dplyr::summarise(mean = mean(width))
```

## 3B. heatmap of osa binding sites vs control

```{r, 3A}

#TODO - functionalize normalizetomatrix call, + common min/max and colorRamp

cnr.mats <- list(mats$yw.rep1, mats$osaGFP.rep1, mats$yw.rep2, mats$osaGFP.rep2) # correct order for subsequent names()
names(cnr.mats) <- cnr.ss$id

common_min = min(unlist(cnr.mats))
common_max = max(unlist(cnr.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

osa.count <- osaPeaks %>%
  data.frame() %>%
  nrow()

osa.count[1] <- as.character(osa.count[1])
  
names(osa.count) <- 'Osa Peak'

groups <- ifelse(osaPeaks$osa.cnr, 'Osa Peak', NA)

cnr.hms <- purrr::map(names(cnr.mats), function(x) {
  if(x == 'osaGFP.rep1'){
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,6))),
                    left_annotation = c(rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(osa.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm'))))) 
  } else {
    EnrichedHeatmap(cnr.mats[[x]],
                    axis_name = c(-2000, 0, 2000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun, 
                    show_heatmap_legend = FALSE,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis = F, ylim = c(0,6)))) 
  }
})
names(cnr.hms) <- names(cnr.mats)

hms.ordered <-cnr.hms$osaGFP.rep1 + cnr.hms$osaGFP.rep2 + cnr.hms$yw.rep1 + cnr.hms$yw.rep2

png('rPlots/3A.png', width = 6, height = 6, units = 'in', res = 300)
hms.ordered
dev.off()
```

## 3B. Osa CnR brDisc browswer shot

```{r, fig.width=4, fig.height=3}

#TODO - best way to generate replicate average signal tracks
#TODO - functionalize givz plotting -- see SLN e93 gof repo - wrote series of functions for browser shot plotting

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
#track.colors <- c("#B8E186","#F1B6DA", "#4DAC26","#D01C8B")
track.colors <- c("grey40","#F1B6DA", "grey20","#D01C8B")

cnr.ss.browser <- cnr.ss %>%
  dplyr::mutate(color = track.colors)

cnr.tracks <- get_cnr_tracks(cnr.ss.browser, ylim = c(0,12))

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/3B.png', width = 4, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, cnr.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```
### 3B-additional CnR browser 
```{r}



track.colors <- c("grey40","#F1B6DA", "grey20","#D01C8B")

cnr.ss.browser <- cnr.ss %>%
  dplyr::mutate(color = track.colors)

cnr.tracks <- get_cnr_tracks(cnr.ss.browser, ylim = c(0,45))

axis <- Gviz::GenomeAxisTrack(scale = 10000)

png('rPlots/3B-espl.png', width = 5, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, cnr.tracks, genetrack), chromosome = seqnames(espl) %>% as.character(), 
                 from = start(espl - 10000),
                 to = end(espl - 40000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()


#####
cnr.ss.browser <- cnr.ss %>%
  dplyr::mutate(color = track.colors)

cnr.tracks <- get_cnr_tracks(cnr.ss.browser, ylim = c(0,50))

axis <- Gviz::GenomeAxisTrack(scale = 10000)

png('rPlots/3B-Dl.png', width = 5, height = 4, res = 300, units = 'in')
Gviz::plotTracks(c(axis, cnr.tracks, genetrack), chromosome = seqnames(Dl) %>% as.character(), 
                 from = start(Dl),
                 to = end(Dl),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```

## 3C. Where are osaGFP binding sites?

```{r, 3C-peak-location}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr | assay == '500bp background') %>%
  dplyr::group_by(assay, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n),
                assay = ifelse(assay == 'cnr', 'Osa-GFP CnR', '500bp Bkg')) %>%
  ggplot(aes(x = '', y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill', color = 'black') +
  geom_text_repel(aes(label= ifelse(pct > 0.07, paste0(sprintf("%1.1f", pct*100),"%"),'')),
                  position=position_fill(vjust=0.5),
                  colour="black", size =5,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_brewer(palette = 'Spectral') +
#  scale_fill_manual(values = cbPalette.ibm) +
  ylab('Fraction') +
  guides(fill = guide_legend(byrow = TRUE)) +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.6, 'cm'),
        legend.spacing.y = unit(0.1, 'cm'),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.text = element_text(size = 12)) +
  facet_grid(~assay)
ggsave('rPlots/3C.png', width = 4, height = 4, units = 'in', dpi = 300)


```

## 3D - AT content within osa peaks

Quantification of the precentage of A and T bases in 3kb windows around
osa peaks. ie. a the fraction of peaks at each position of a 3kb window
that have A or T bases. Suggests there is lower A/T content within osa
binding sites.

Sanity check that summit coordinates don't change this result, and they
don't. Summit coords are likely off center from the narrowpeak
coordinate center, so maybe there's a difference in AT content around
summits. But that doesn't appear to be true, still saw a drop in AT
content.

```{r}
bw_list <- c('BigWig/osaGFP-3LW-wing-aGFP-CnR-sup-rep1_dm6_trim_q5_dupsRemoved_allFrags_rpgcNorm_zNorm.bw',
             'BigWig/osaGFP-3LW-wing-aGFP-CnR-sup-rep2_dm6_trim_q5_dupsRemoved_allFrags_rpgcNorm_zNorm.bw')
  
bws <- rtracklayer::BigWigFileList(bw_list)

names(bws) <- c('rep1','rep2')

zmatrix <- purrr::map(names(bws), function(x) {
  signal::get_average_signal(bws[[x]], osaPeaks %>% resize(3000,'center')) %>% dplyr::mutate(id = x)
  }) %>%
  dplyr::bind_rows(.) %>%
  dplyr::group_by(position) %>%
  dplyr::summarise(meanZ = mean(value.mean)) %>%
  dplyr::mutate(name = position - 1500,
                fracZ = meanZ/max(meanZ))

##

osaBound.seq = osaPeaks %>% 
  resize(3000, "center") %>%
  memes::get_sequence(dm6)

osaBound.seq.df <- osaBound.seq %>%
  data.frame() 

colnames(osaBound.seq.df) <- 'seqs'

osaBound.seq.split <- stringr::str_split(osaBound.seq.df$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(osaBound.seq.split) <- as.character(c(1:3000))

osaBound.df <- dplyr::bind_cols(osaBound.seq.split, data.frame(osaPeaks))

osaBound.df %>%
  tidyr::pivot_longer(c(1:3000)) %>%
  dplyr::mutate(name = as.integer(name),
                name = name - 1500) %>%
  #dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
  #              AT = ifelse(value == 'T' | value == 'A', T, F),
  #              name = as.integer(name)) 
  dplyr::select(peak, name, value) %>%
  dplyr::group_by(name, value) %>%
  dplyr::tally() %>%
  dplyr::filter(value != 'N') %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  dplyr::left_join(zmatrix,by = 'name') %>%
  dplyr::filter(value %in% c('A','T')) %>%
  ggplot(aes(name, freq, color = value)) +
  geom_point(alpha = 0.1) +
  scale_color_manual(values = c('#D81B60','#1E88E5'), name = '') +
  geom_smooth(method = 'loess', span = 0.1) +
  geom_line(aes(y = fracZ/2, x = name), color = 'black') +
  #geom_line(aes(y = (meanZ/20)+0.1, x = name)) +
#  ylim(c(0.15,0.35)) +
  scale_y_continuous(name = 'AT Frequency',, 
                     sec.axis = sec_axis(~.*2, name = 'Fraction of Max Osa zScore')) + 
  ylab('Fraction of Osa Peaks') +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        legend.position = 'left')
ggsave('rPlots/3-draft-ATcontent.png', dpi = 300, width = 4, height = 3)
  #geom_bar(stat = 'identity', position = 'fill')

#####
```

## 3D. How does osa binding correlate with wt faire data?

```{r 3LW-FAIRE-in-OsaPeaks}
#mats <- purrr::map(bws, ~{
#  normalizeToMatrix(., osaPeaks, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
#})
#names(mats) <- names(bws)
osaPeaks.enhancers <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  GRanges()

#fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
wt.3lw.mat <- normalizeToMatrix(bws$wt_3LW, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.1.mat <- normalizeToMatrix(bws$osaGFP.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.2.mat <- normalizeToMatrix(bws$osaGFP.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
fig3d.mats <- list(wt.3lw.mat, osa.1.mat, osa.2.mat)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks.enhancers$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks.enhancers %>%
  data.frame() %>%
  dplyr::group_by(osa.cnr, WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm')))),
                                                 #rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'right'))))
  

png('rPlots/3D.png', width = 5, heigh = 6, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, heatmap_legend_side = "bottom", merge_legend = T)
dev.off()


```
### control version
#### control all peaks
```{r}
osaPeaks.enhancers <- peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(osa.cnr & yw.cnr | yw.cnr & !osa.cnr) %>%
#  dplyr::filter(osa.cnr & !yw.cnr & log2FoldChange >= 0) %>%
#  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  GRanges() %>%
  resize(width = 1, fix = 'center')
#osaPeaks.enhancers
#fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
wt.3lw.mat <- normalizeToMatrix(bws$wt_3LW, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.1.mat <- normalizeToMatrix(bws$osaGFP.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.2.mat <- normalizeToMatrix(bws$osaGFP.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.1.mat <- normalizeToMatrix(bws$yw.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.2.mat <- normalizeToMatrix(bws$yw.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
fig3d.mats <- list(wt.3lw.mat, osa.1.mat, osa.2.mat, yw.1.mat,yw.2.mat)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2', 'yw rep1', 'yw rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks.enhancers$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks.enhancers %>%
  data.frame() %>%
  dplyr::group_by(WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm')))),
                                                 #rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,8), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,8), axis = F))) +
    EnrichedHeatmap(fig3d.mats$`yw rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'yw.1',
                  col = col_fun,
                  name = 'yw zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,8), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`yw rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'yw.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,8), axis_param = list(side = 'right'))))

png('rPlots/3-control-allPeaks.png', width = 5, heigh = 6, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, heatmap_legend_side = "bottom", merge_legend = T)
dev.off()


```
#### shared peaks
```{r}
osaPeaks.enhancers <- peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(osa.cnr & yw.cnr) %>%
#  dplyr::filter(osa.cnr & !yw.cnr & log2FoldChange >= 0) %>%
#  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  GRanges() %>%
  resize(width = 1, fix = 'center')
#osaPeaks.enhancers
#fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
wt.3lw.mat <- normalizeToMatrix(bws$wt_3LW, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.1.mat <- normalizeToMatrix(bws$osaGFP.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.2.mat <- normalizeToMatrix(bws$osaGFP.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.1.mat <- normalizeToMatrix(bws$yw.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.2.mat <- normalizeToMatrix(bws$yw.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
fig3d.mats <- list(wt.3lw.mat, osa.1.mat, osa.2.mat, yw.1.mat,yw.2.mat)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2', 'yw rep1', 'yw rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks.enhancers$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks.enhancers %>%
  data.frame() %>%
  dplyr::group_by(WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm')))),
                                                 #rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,6), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,11), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,11), axis = F))) +
    EnrichedHeatmap(fig3d.mats$`yw rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'yw.1',
                  col = col_fun,
                  name = 'yw zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,11), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`yw rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'yw.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,11), axis_param = list(side = 'right'))))

png('rPlots/3-sharedPeaks.png', width = 5, heigh = 6, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, heatmap_legend_side = "bottom", merge_legend = T)
dev.off()


```
#### control specific peaks
```{r}
osaPeaks.enhancers <- peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(yw.cnr & !osa.cnr) %>%
#  dplyr::filter(osa.cnr & !yw.cnr & log2FoldChange >= 0) %>%
#  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  GRanges() %>%
  resize(width = 1, fix = 'center')
#osaPeaks.enhancers
#fig3d.mats <- list(mats$wt_3LW, mats$osaGFP.rep1, mats$osaGFP.rep2)
wt.3lw.mat <- normalizeToMatrix(bws$wt_3LW, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.1.mat <- normalizeToMatrix(bws$osaGFP.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
osa.2.mat <- normalizeToMatrix(bws$osaGFP.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.1.mat <- normalizeToMatrix(bws$yw.rep1, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
yw.2.mat <- normalizeToMatrix(bws$yw.rep2, osaPeaks.enhancers, value_column = 'score', keep = c(0.01, 0.99), extend = 2000)
fig3d.mats <- list(wt.3lw.mat, osa.1.mat, osa.2.mat, yw.1.mat,yw.2.mat)
names(fig3d.mats) <- c('WT_3LW', 'Osa Rep1', 'Osa Rep2', 'yw rep1', 'yw rep2')
  
common_min = min(unlist(fig3d.mats))
common_max = max(unlist(fig3d.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- ifelse(osaPeaks.enhancers$WT.3LW, "Open", "Closed")

lgd = Legend(at = c("closed", "open"), legend_gp = gpar(fill = c('#fc5d63','#87d674')))

l3.count <- osaPeaks.enhancers %>%
  data.frame() %>%
  dplyr::group_by(WT.3LW) %>%
  dplyr::count() %>%
  dplyr::mutate(WT_FAIRE_3LW = ifelse(WT.3LW, 'Open','Closed'),
                n = as.character(n)) %>%
  .$n

names(l3.count) = unique(groups)

fig3d.hms <- EnrichedHeatmap(fig3d.mats$WT_3LW,
                             axis_name = c(-2,0,+2),
                             pos_line = F,
                             column_title = '3LW-FAIRE',
                             col = viridis.hex,
                             width = unit(1.5, 'in'), 
                             left_annotation = c(rowAnnotation(textbox = anno_textbox(groups, 
                                                                                      as.list(l3.count), 
                                                                                      side = 'left',
                                                                                      background_gp = gpar(fill = NA, col = NA),
                                                                                      gp = gpar(col = 'black'), 
                                                                                      padding = unit(0, 'mm')))),
                                                 #rowAnnotation(groups = anno_block(gp = gpar(fill = c('#fc5d63','#87d674')), show_name = F))), 
                             top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,4.5), axis_param = list(side = 'left'))),  
                             row_title = NULL, 
                             name = 'FAIRE zScore') +
  EnrichedHeatmap(fig3d.mats$`Osa Rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'Osa.1',
                  col = col_fun,
                  name = 'Osa zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`Osa Rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'Osa.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
    EnrichedHeatmap(fig3d.mats$`yw rep1`,
                  axis_name = c(-2,0,+2),
                  pos_line = F,
                  column_title = 'yw.1',
                  col = col_fun,
                  name = 'yw zScore',
                  width = unit(0.5,'in'),
                  top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis = F))) +
  EnrichedHeatmap(fig3d.mats$`yw rep2`,
                   axis_name = c(-2,0,+2),
                   pos_line = F,
                   column_title = 'yw.2',
                   col = col_fun,
                   show_heatmap_legend = F, 
                   width = unit(0.5,'in'),
                   top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = c('#fc5d63','#87d674')), ylim = c(0,7), axis_param = list(side = 'right'))))

png('rPlots/3-control-specificPeaks.png', width = 5, heigh = 6, units = 'in', res = 300)
ComplexHeatmap::draw(fig3d.hms, row_split = groups, annotation_legend_list = list(lgd), row_title = NULL, heatmap_legend_side = "bottom", merge_legend = T)
dev.off()


```

### scatterplot
```{r}
peaks$cnr.df %>%
  dplyr::filter(osa.cnr & yw.cnr | !osa.cnr & yw.cnr) %>%
  #dplyr::filter(faire.3LW.zScore <= 1) %>%
 # dplyr::mutate(neglogp = -log(cnr.padj)) %>%
  ggplot(aes(faire.3LW.zScore, yw.rep2.zScore, )) +
  geom_point() +
  ylim(c(0,10))
```

## 3E. Fraction of Osa binding sites withing WT 3LW-24h FAIRE categories
tested with requiring 3lw and/or 24h peak calls 
doesnt dramatically chage osa binding percentage, but doesn reduce closing category aboslute number by ~50%, static category goes down some too
```{r, 3E}
#TODO - combine fraction barplots and heatmaps into single figure - cowplots? grid?
fairePeaks.wt.test <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h | !WT.3LW & !WT.6h & WT.24h & !WT.44h) %>%
  GRanges()

faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.wt.test, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.wt.test, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat, faire.24h.mat))
common_max = max(c(faire.3lw.mat, faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.wt.test %>%
  data.frame() %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h | !WT.3LW & !WT.6h & WT.24h & !WT.44h) %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::count() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')

groups <- fairePeaks.wt.test %>%
  data.frame() %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h | !WT.3LW & !WT.6h & WT.24h & !WT.44h) %>%
  .$faireCat.3LW_24h

lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                      left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 



png('rPlots/3E.png', width = 4, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hm, annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'left', )
dev.off() 
 
fairePeaks.wt.test %>%
  data.frame() %>%
#peaks %>% 
 # dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h | !WT.3LW & !WT.6h & WT.24h & !WT.44h) %>%
#  dplyr::filter(WT.3LW | WT.24h) %>%
  dplyr::group_by(faireCat.3LW_24h, osa.cnr, yw.cnr) %>%
  dplyr::count() %>%
  na.omit() %>%
  dplyr::mutate(Osa.Bound = ifelse(osa.cnr & !yw.cnr, T, F)) %>%
  dplyr::ungroup() %>%
  dplyr::select(faireCat.3LW_24h, Osa.Bound, n) %>%
  reshape2::melt(id = c("Osa.Bound","n"), variable.name = 'stage', value.name = 'change') %>%
  dplyr::group_by(stage, change, Osa.Bound) %>%
  dplyr::summarise(peaks = sum(n)) %>%
  dplyr::mutate(fraction = peaks / sum(peaks),
                change = factor(change, levels = c('static','opening','closing'))) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()

```

## Drafts
### scatter of osa vs control 
```{r}
peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(assay == 'cnr' & osa.cnr | yw.cnr) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(osaGFP = mean(c(osaGFP.rep1.zScore, osaGFP.rep2.zScore)),
                control = mean(c(yw.rep1.zScore, yw.rep2.zScore))) %>% 
  dplyr::select(peak, cnr.peakCat, osaGFP, control) %>%
#  tidyr::pivot_longer(c(osaGFP, control)) %>%
  ggplot(aes(osaGFP, control, color = cnr.peakCat)) +
    geom_point() +
  xlim(c(0,25)) +
  ylim(c(0,25))


peaks$cnr.df %>%
  data.frame() %>%
  dplyr::filter(assay == 'cnr' & osa.cnr | yw.cnr) %>%
  dplyr::mutate(neglogP = -log10(padj)) %>%
#  tidyr::pivot_longer(c(osaGFP, control)) %>%
  ggplot(aes(log2FoldChange, neglogP, color = cnr.peakCat)) +
    geom_point(alpha = 0.5) 
```

### clustering faire data
reproducing the clustered FAIRE heatmap that SLN made for 2020 paper
he used mean cpm from rsubread counts
here I'm just just pooled zNorm signal within peaks - not sure if one method is better than the other
the clusters look very similar to what we previously published.

TODO - what is the osa binding overlap within each cluster?
  - get peak ID from each cluster than match the peaks df and plot fraction that are osa.cnr True

```{r}

tidymat <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron')) %>%
  dplyr::select(peak, dplyr::matches("faire.\\d{1,2}h.zScore|faire.3LW.zScore")) %>%
  tidyr::pivot_longer(!peak) %>%
#  dplyr::group_by(peak) %>%
  dplyr::mutate(stage = stringr::str_split_fixed(name,'\\.', n = 3)[,2], 
                #stage = stringr::str_remove(stage, 'h'), 
                #stage = ifelse(stage == '3LW', -6, stage),
                stage = factor(stage, levels = c('3LW', '6h', '18h','24h','36h','44h'))) %>%
  dplyr::group_by(peak) %>%
  dplyr::mutate(value = value +1 , 
                frac = value/max(value),
                peak = as.character(peak)) %>%
  tidyr::pivot_wider(id_cols = peak, names_from = stage, values_from = frac) %>%
  tibble::column_to_rownames('peak') %>%
  as.matrix()

# based of nystrom
set.seed(1230412)
k8 <- kmeans(tidymat, 8, nstart = 24, iter.max = 15)
figure_col_fun <- circlize::colorRamp2(breaks = c(0,0.25,0.5,0.75,1), colors = viridis::viridis(5))

k_df <- k8$cluster %>% 
  data.frame %>% 
  dplyr::rename(k = ".") %>% 
  tibble::rownames_to_column("id")  
#  dplyr::mutate(new_k1 = dplyr::case_when(k == 1 ~ 8,
#                                          k == 2 ~ 7,
#                                          k == 3 ~ 5,
#                                          k == 4 ~ 2,
#                                          k == 5 ~ 6,
#                                          k == 6 ~ 4,
#                                          k == 7 ~ 3,
#                                          k == 8 ~ 1)) %>% 
#  dplyr::rename(old_k = k,
#                k = new_k1)

png('rPlots/FAIRE-cluster-heatmap.png', width = 5, height = 4, res = 300, units = 'in')
ComplexHeatmap::Heatmap(tidymat, split = k_df$k, cluster_columns = F, show_row_dend = F, col = figure_col_fun, gap = unit(0.5, "lines"),
                        show_row_names = F)
dev.off()
```
```{r}
# annotate all WT FAIRE peaks with flanking gene info - returns more than 3 genes because of shared distances
fairePeaks.anno <- ChIPseeker::annotatePeak(fairePeaks.wt, tssRegion = c(-100,500), TxDb = dm6.TxDb, level = 'gene', addFlankGeneInfo = T, flankDistance = 5000) %>%
  ChIPseeker::as.GRanges() %>%
  data.frame() %>%
  dplyr::mutate(Geneid_flank = stringr::str_split(flank_geneIds, ';'), 
                Gene_distance = stringr::str_split(flank_gene_distances, ';')) %>%
#  tidyr::unnest(Geneid) %>%
  #tidyr::unnest(Gene_distance) %>% 
#  dplyr::mutate(Geneid = ifelse(is.na(Geneid), geneId, Geneid)) %>% # in case where there are no genes withing 10kb, take the nearest gene annotation
  dplyr::group_by(peak) %>%
#  dplyr::slice_min(Gene_distance, n = 4) %>%
  dplyr::rename(Geneid = geneId) %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid')
```

```{r}

## merge in clusters to peak list and then plot line and pie charts

cluster.df <- data.frame(k8$cluster) %>%
  tibble::rownames_to_column('peak') %>%
  dplyr::mutate(peak = as.integer(peak)) %>% 
  dplyr::left_join(data.frame(fairePeaks.wt), ., by = 'peak') %>% # join cluster info to the wt FAIRE peak df
  dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron')) %>% # filter out to just "distal regulatory"
  dplyr::mutate(osabound = ifelse(osa.cnr & !yw.cnr, T, F),
                k8.cluster = factor(k8.cluster, levels = c('8', '3', '4','5','1','6','2','7'))) %>%
  dplyr::rename('Geneid' = 'geneId') %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid')

cluster.df %>% dplyr::filter(k8.cluster ==8) #%>% dplyr::filter(osa.cnr & !yw.cnr)
```


```{r}
purrr::map(unique(cluster.df$k8.cluster), function(x) {
  name <- paste0("cluster-",x)
  
 piechart <- cluster.df %>%
    dplyr::filter(k8.cluster == x) %>%
    dplyr::distinct(peak, .keep_all = T) %>%
    dplyr::group_by(osabound) %>%
    dplyr::tally() %>% 
    dplyr::mutate(frac = n/sum(n),
                  percent = paste0(100*round(frac, 2),"%")) %>%
    ggplot(aes(0,frac, fill = osabound, label = percent)) +
    geom_bar(stat = 'identity', position = position_stack(vjust = 0.5)) + 
    geom_text(position = position_stack(vjust = 0.5)) +
    scale_x_continuous(expand = c(0,0)) +
    coord_polar(theta = 'y') + 
    scale_fill_manual(values = c('grey80', '#e667d3')) + 
    theme(axis.text = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          panel.background = element_blank(),
          strip.background = element_blank(),
          strip.text = element_blank(),
          legend.position = 'none')  
  
  lineplot <- cluster.df %>%
    dplyr::filter(k8.cluster == x) %>%
    dplyr::distinct(peak, .keep_all = T) %>%
     dplyr::select(peak, k8.cluster, dplyr::matches("faire.\\d{1,2}h.zScore|faire.3LW.zScore")) %>%
     tidyr::pivot_longer(dplyr::matches("faire.\\d{1,2}h.zScore|faire.3LW.zScore")) %>%
     dplyr::mutate(stage = stringr::str_split_fixed(name,'\\.', n = 3)[,2],
                   stage = factor(stage, levels = c('3LW', '6h', '18h','24h','36h','44h')),
                   k8.cluster = factor(k8.cluster, levels = c('8', '3', '4','5','1','6','2','7'))) %>%
     dplyr::group_by(peak) %>%
     dplyr::mutate(frac = value/max(value)) %>%
     dplyr::group_by(k8.cluster, stage) %>%
     dplyr::summarise(mean = mean(frac),
                      sd = sd(frac) ,
                      n = dplyr::n()) %>%
     ggplot(aes(stage, mean, group = k8.cluster)) +
     geom_ribbon(aes(y = mean, ymin = mean-sd, ymax = mean+sd), alpha = 0.3) +
     geom_line(aes(color = k8.cluster), size = 2) +
     ylim(c(-0.1,1.1)) +
     theme(legend.position = 'none')
  
  
#rna <-   cluster.df %>%
#  dplyr::filter(!grepl('RNA',SYMBOL)) %>% # filter out stuff like lncRNA, tRNA, snoRNA etc. 
#   dplyr::filter(k8.cluster == x) %>% 
#    dplyr::select(Geneid,k8.cluster,osabound,c(68:79)) %>% #TODO stop indexing like this!
#    tidyr::pivot_longer(c(4:15)) %>%    
#    dplyr::distinct() %>% #should drop any genes represented >1 per cluster... 
#  #TODO -- good idea to filter on distinct() - quick check this doesn't look like it dramatically changes the plots, check numbers. 
#    dplyr::mutate(stage = stringr::str_split_fixed(name, '_', n = 2)[,1],
#                  stage = stringr::str_remove(stage, 'X'),
#                  stage = factor(stage, levels = c('L3', '6h', '18h','24h','36h','44h')),
#                  rep = stringr::str_split_fixed(name, '_', n = 2)[,2]) %>%
#    dplyr::group_by(Geneid, stage, k8.cluster, osabound) %>%
#    dplyr::summarise(rep_mean = mean(value)+0.001) %>% # average rnaseq replicates
#    dplyr::group_by(Geneid, k8.cluster) %>%
#   dplyr::filter(any(rep_mean >= 2)) %>% # drop genes with little to no expression
#    dplyr::group_by(k8.cluster,stage, osabound) %>% 
#    dplyr::summarise(mean = mean(rep_mean), 
#                     sd = sd(rep_mean)) %>% #average all the fraction of max together per stage per cluster
#   dplyr::group_by(k8.cluster, osabound) %>%
#   dplyr::mutate(frac = mean/max(mean)) %>%
#    ggplot(aes(stage, frac, group = osabound,color = osabound) ) +
#    geom_line(size = 2) +
#  ylim(c(0,1.0)) + 
#   scale_color_manual(values = c('grey50','magenta')) +
#    theme(legend.position = 'none') 
   

 ggsave(plot = piechart, filename = paste0('rPlots/pieChart-',name,'.png'), width = 1, height = 1 , dpi = 300) 
 ggsave(plot = lineplot, filename = paste0('rPlots/linePolot-',name,'.png'), width = 3, height = 3 , dpi = 300) 
# ggsave(plot = rna, filename = paste0('rPlots/rna-',name,'.png'), width = 3, height = 3 , dpi = 300) 
    
})

  #plot average of faire fract max per cluster - simple abstraction for figure
cluster.df %>%
#    dplyr::filter(k8.cluster == x) %>%
    dplyr::distinct(peak, .keep_all = T) %>%
     dplyr::select(peak, k8.cluster, dplyr::matches("faire.\\d{1,2}h.zScore|faire.3LW.zScore")) %>%
     tidyr::pivot_longer(dplyr::matches("faire.\\d{1,2}h.zScore|faire.3LW.zScore")) %>%
     dplyr::mutate(stage = stringr::str_split_fixed(name,'\\.', n = 3)[,2],
                   stage = factor(stage, levels = c('3LW', '6h', '18h','24h','36h','44h')),
                   k8.cluster = factor(k8.cluster, levels = c('8', '3', '4','5','1','6','2','7'))) %>%
     dplyr::group_by(peak) %>%
     dplyr::mutate(frac = value/max(value)) %>%
     dplyr::group_by(k8.cluster, stage) %>%
     dplyr::summarise(mean = mean(frac),
                      sd = sd(frac) ,
                      n = dplyr::n()) %>%
  dplyr::mutate(k8.cluster = factor(k8.cluster, levels = c('4','8','5','2','3','6','7','1'))) %>%
     ggplot(aes(stage, k8.cluster, fill = mean ))+
     geom_raster() +
  scale_fill_viridis_c() +
  theme(axis.text = element_text(size = 15),
        axis.title = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'))
ggsave('rPlots/3I-FAIRE-cluster-simple.png', width = 4, height = 3, dpi = 300)


```

#### ame of clusters
```{r}
#TODO functionalize to process all
cluster.1 <- data.frame(k8$cluster) %>%
  tibble::rownames_to_column('peak') %>%
  dplyr::mutate(peak = as.integer(peak)) %>%
  dplyr::left_join(peaks, ., by = 'peak') %>%
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron') & k8.cluster == 8 & osa.cnr & !yw.cnr) %>%
  memes::get_sequence(dm6)

cluster.control <- data.frame(k8$cluster) %>%
  tibble::rownames_to_column('peak') %>%
  dplyr::mutate(peak = as.integer(peak)) %>%
  dplyr::left_join(peaks, ., by = 'peak') %>%
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron') & k8.cluster != 8 & osa.cnr & !yw.cnr) %>%
  memes::get_sequence(dm6)

ame <- memes::runAme(cluster.1, control = cluster.control) %>% 
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 
#streme <- memes::runStreme(cluster.1, control = cluster.control)  
#memes::runTomTom(streme)


ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:10) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 
ggsave('rPlots/cluster-8_ame.png', width = 3, height = 3, dpi = 300)
```

#### go of clusters
In progress, there doesn't look like there's particularly strong enrichment within cluster 1, depends on how the analysis is controlled though. 
When subset by osa bound, which seems most appropriate to me, its hard to find enrichment unless a very broad control is set, in which case terms tend to be mainly associated with appendage/wing dev and metamorphosis - nothing terribly specific. 

cluster 1 osaBound vs all clusters - no enrichment
cluster 1 osaBound vs all "on" genes - wing dev, appendage, metamorph, etc. 
cluster 7 osaBound vs all clusters - anterior/posterior specification, segmentation, imag disc dev. 

wiki pathway analysis -- 
osa bound cluster 1 enriched for notch signalling 
osa bound cluster 4 enriched for wnt signalling

```{r}

cluster <- cluster.df %>%
  dplyr::filter(osa.cnr & !yw.cnr & !grepl('RNA', SYMBOL)) %>%
  dplyr::filter(k8.cluster == 1 ) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 



control <- ON.genes.validated %>%
  dplyr::ungroup() %>%
#  dplyr::filter(osa.cnr & !yw.cnr) %>%
#  dplyr::filter(k8.cluster == 1 | k8.cluster == 4 ) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 

#goEnrich(cluster, control, 'test')

#test <- cluster.df %>% dplyr::ungroup() %>% dplyr::filter(k8.cluster == 1 & osabound) %>% GRanges() + 1000
#test

#test2 <- ChIPseeker::seq2gene(test, tssRegion = c(-100,500), TxDb = dm6.TxDb, flankDistance = 5000) %>% data.frame('Geneid' = .)

entrez <- clusterProfiler::bitr(cluster$Geneid, 'FLYBASE', 'ENTREZID', OrgDb = org.Dm.eg.db)$ENTREZID

clusterProfiler::enrichWP(entrez, organism = 'Drosophila melanogaster')

```
##### scratch
```{r}
list <- cluster.df %>%
  dplyr::filter(k8.cluster == 7 & osa.cnr & !yw.cnr) %>%
  dplyr::select(Geneid, matches('L3_rep\\d'), matches('\\d{1,2}h_rep\\d')) %>%
  dplyr::distinct()

symbol <- clusterProfiler::bitr(list$Geneid, 'FLYBASE', 'SYMBOL', OrgDb = org.Dm.eg.db) %>% 
  dplyr::rename(Geneid = FLYBASE)
symbol

cluster1.mat <- list %>%
  dplyr::rowwise() %>%
  dplyr::left_join(., symbol, by = 'Geneid') %>%
  dplyr::select(!Geneid) %>%
  tidyr::pivot_longer(!SYMBOL) %>%
  dplyr::mutate(stage = stringr::str_split_fixed(name,'_', n = 2)[,1],
                #stage = stringr::str_remove(stage, 'h'), 
                #stage = ifelse(stage == '3LW', -6, stage),
                stage = factor(stage, levels = c('L3', '6h', '18h','24h','36h','44h'))) %>%
  dplyr::group_by(SYMBOL, stage) %>%
  dplyr::summarise(mean = mean(value)) %>%
  dplyr::group_by(SYMBOL) %>%
  dplyr::mutate(frac = mean / max(mean),
                frac = ifelse(is.na(frac), 0, frac)) %>%
  dplyr::select(!mean) %>%
  tidyr::pivot_wider(id_cols = SYMBOL, names_from = stage, values_from = frac) %>%
  tibble::column_to_rownames('SYMBOL') %>%
  as.matrix()
  
cluster1.mat
set.seed(1230412)
k5 <- kmeans(cluster1.mat, 8, nstart = 24, iter.max = 15)  
  
figure_col_fun <- circlize::colorRamp2(breaks = c(0,0.25,0.5,0.75,1), colors = viridis::viridis(5))

k5_df <- k5$cluster %>% 
  data.frame %>% 
  dplyr::rename(k = ".") %>% 
  tibble::rownames_to_column("id")# %>% 
#  dplyr::mutate(new_k1 = dplyr::case_when(k == 1 ~ 8,
#                                          k == 2 ~ 7,
#                                          k == 3 ~ 5,
#                                          k == 4 ~ 2,
#                                          k == 5 ~ 6,
#                                          k == 6 ~ 4,
#                                          k == 7 ~ 3,
#                                          k == 8 ~ 1)) %>% 
#  dplyr::rename(old_k = k,
#                k = new_k1)


ComplexHeatmap::Heatmap(cluster1.mat, split = k5_df$k, cluster_columns = F, show_row_dend = F, col = figure_col_fun, gap = unit(0.5, "lines"),
                        show_row_names = T)
```


```{r}
k5_df %>% dplyr::filter(k == 1)
```


### binding around tss
```{r}

osaPeaks.promoter <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(-1000 < distanceToTSS & distanceToTSS > 500 ) %>%
  GRanges() %>% 
  resize(1000, 'center')


test <- subsetByOverlaps(dm6.promoters, osaPeaks.promoter, maxgap = 10000, minoverlap = 0) 

matrix1 <- EnrichedHeatmap::normalizeToMatrix(c(bws$osaGFP.rep1, bws$osaGFP.rep2), test, value_column = 'score')
#matrix2 <- EnrichedHeatmap::normalizeToMatrix(bws$osaGFP.rep2, dm6.promoters, value_column = 'score')

EnrichedHeatmap::EnrichedHeatmap(matrix1)

```

### enhancer binding
what about defining "enhancers" as 3LW faire peak calls?
or signal?
```{r}
osaPeaks %>%
  data.frame() %>%
  #dplyr::filter(WT.3LW) %>%
  dplyr::group_by(WT.3LW, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n),
                WT.3LW = ifelse(WT.3LW,'Osa Bound - 3LW FAIRE Peak','Closed')) %>%
#                assay = ifelse(assay == 'cnr', 'Osa-GFP CnR', '1kb Bkg')) %>%
   ggplot(aes(x = '', y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill', color = 'black') +
  geom_text_repel(aes(label= ifelse(pct > 0.05, paste0(sprintf("%1.1f", pct*100),"%"),'')),
                  position=position_fill(vjust=0.5),
                  colour="black", size =5,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_brewer(palette = 'Spectral') +
#  scale_fill_manual(values = cbPalette.ibm) +
  ylab('Fraction') +
  guides(fill = guide_legend(byrow = TRUE)) +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.6, 'cm'),
        legend.spacing.y = unit(0.1, 'cm'),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.text = element_text(size = 12)) +
  facet_grid(~WT.3LW)
ggsave('rPlots/3-draft-osa_annotation_3LW_fairePeaks.png', width = 3, height = 5, dpi = 300)
```

### location of osa closing sites
```{r}
peaks %>%
  dplyr::filter(assay == 'faire'  & experiment == 'WT FAIRE Wing Timecourse') %>%
  #dplyr::filter(assay == '' &  osa.cnr & !yw.cnr | assay == '1kb background') %>%
  dplyr::filter(osa.cnr & !yw.cnr & faireCat.3LW_24h != 'opening') %>%
  dplyr::mutate(cnr.peakCat = ifelse(osa.cnr & !yw.cnr, 'Osa Bound', 'Osa Unbound')) %>%
  dplyr::group_by(faireCat.3LW_24h, cnr.peakCat, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
#                assay = ifelse(assay == 'cnr', 'Osa-GFP CnR', '1kb Bkg')) %>%
  ggplot(aes(x = '', y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill', color = 'black') +
  geom_text_repel(aes(label= ifelse(pct > 0.05, paste0(sprintf("%1.1f", pct*100),"%"),'')),
                  position=position_fill(vjust=0.5),
                  colour="black", size =5,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_brewer(palette = 'Spectral') +
#  scale_fill_manual(values = cbPalette.ibm) +
  ylab('Fraction') +
  guides(fill = guide_legend(byrow = TRUE)) +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.6, 'cm'),
        legend.spacing.y = unit(0.1, 'cm'),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.text = element_text(size = 12)) +
  facet_grid(cnr.peakCat ~ faireCat.3LW_24h)
ggsave('rPlots/3-draft-peakCat-faireCat.png', width = 4, height = 4, dpi = 300)
```

### nearest gene - other methods

osaBound / unBound -- 

closing.osaBound / osaBound -- morphogenesis, epithelial tube, wing dev
static.osaBound / osaBound -- very similar
closing.osaBound / static.osaBound -- no enrichment

changing to use fairepeaks instead of osa

summarizing / grouping terms using `rrvgo` 
*TODO* 
 - test analysis with other categories 
 - finalize rnaseq differential cat 
   - switch to deseq?
 
down reg genes that are near closing osa bound sites 
and near static osa bound sites are enriched for terms associated with neurogenesis.

There's no enrichment of terms between these two groups. 

suggests that osa bound sites might be near neuro genes in general?

--- old ---- 1. Identify index for nearest osa peak for all dm6 genes 2.
join the dm6 gene ranges object with peak coordinates using osa peak
index 3. calculate distance between gene start and osa peak center 4.
select the top 3 closest genes for each osa peak 5. join with wing rna
seq data and calculate log2fold change of expression between L3 and 24h
6. obtain list of gene id for all genes near osa peaks that overlap with
faire peaks that close from l3 to 24h and have reduced expression in the
same period

Logic -- sites with reduced accessibility may be more likely to have a
regulatory connection to nearby genes with reduced expression

result -- there are 89 downreg genes near osa bound closing sites these
genes are enriched for BP GO terms associated with: "cell fate
specification", "Dendrite development," "wing morphogenesis," and
"transcriptional regulation - positive and negative"

\*\* trying with control for all down regulated genes still get
dendrite, but it drops some on the list \##### GO


```{r}
goData <- GOSemSim::godata(org.Dm.eg.db, ont='BP', keytype = 'FLYBASE')


goEnrich <- function(geneList, universe, name) {  
#  go <- select(org.Dm.eg.db, keys = geneList$Geneid, columns = c('SYMBOL','GO','ENTREZID'), keytype = 'FLYBASE')
#  go.universe <- select(org.Dm.eg.db, keys = universe$Geneid, columns = c('SYMBOL','GO','ENTREZID'), keytype = 'FLYBASE')
  ggo <- clusterProfiler::enrichGO(#gene = go$ENTREZID,
                                   gene = geneList$Geneid,
                                   universe = universe$Geneid,
                                   OrgDb = org.Dm.eg.db,
                                   keyType = 'FLYBASE',
                                   ont = 'BP',
                                   readable = T,
                                   pAdjustMethod = 'BH',
                                   pvalueCutoff = 0.01,
                                   qvalueCutoff = 0.05)
  
  ggo.df <- data.frame(ggo)

  simMatrix <- rrvgo::calculateSimMatrix(ggo.df$ID,
                          orgdb='org.Dm.eg.db',
                          ont='BP',
                          method='Rel', keytype = 'ENTREZID', semdata = goData)
  scores <- setNames(-log10(ggo.df$qvalue), ggo.df$ID)
  reducedTerms <- rrvgo::reduceSimMatrix(simMatrix, scores, threshold = 0.5, orgdb = org.Dm.eg.db)
  
 plot <- reducedTerms %>%
    dplyr::group_by(cluster, parentTerm) %>%
    dplyr::mutate(neg_log10_qval = mean(score)) %>%
    dplyr::ungroup() %>%
    dplyr::select(parentTerm, neg_log10_qval) %>%
    dplyr::group_by(parentTerm, neg_log10_qval) %>%
    dplyr::summarise() %>%
    dplyr::ungroup() %>%
    dplyr::slice_max(neg_log10_qval, n = 10) %>%
    dplyr::mutate(parentTerm = factor(parentTerm),
                  parentTerm = forcats::fct_reorder(parentTerm, neg_log10_qval)) %>%
    ggplot(aes('', parentTerm, fill = neg_log10_qval)) +
    geom_tile() +
    scale_fill_viridis_b(name = 'avg\n-log10(qval)') + 
    ggtitle(name) +
    theme(axis.text = element_text(size = 12),
          axis.title = element_blank(),
          axis.ticks.x = element_blank(),
          legend.text = element_text(size = 12),
          legend.title = element_text(size = 12))
    return(ggo.df)
}
```

```{r}
symbol.lookup <- clusterProfiler::bitr(, fromType = 'FLYBASE', toType = 'SYMBOL', OrgDb = org.Dm.eg.db)

data.frame(dm6.txdb)

fairePeaks <- peaks %>% 
  dplyr::filter(assay == 'faire'  & experiment == 'WT FAIRE Wing Timecourse')


anno.test <- ChIPseeker::annotatePeak(GRanges(fairePeaks),tssRegion = c(-100,500), TxDb = dm6.TxDb, addFlankGeneInfo = T, level = 'gene')


anno.test.df <- ChIPseeker::as.GRanges(anno.test) %>% data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron') & osa.cnr & !yw.cnr) 

anno.test.df.unnest <- anno.test.df %>%
  dplyr::mutate(flank_genes = stringr::str_split(flank_geneIds, pattern = ';'),
                flank_distance = stringr::str_split(flank_gene_distances, pattern = ';')) %>%
  tidyr::unnest(c(flank_genes, flank_distance)) %>%
  dplyr::distinct() %>% 
  dplyr::filter(!is.na(flank_genes)) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(SYMBOL = clusterProfiler::bitr(flank_genes, 'FLYBASE','SYMBOL',org.Dm.eg.db))

  dplyr::filter(!grepl('lncRNA',SYMBOL))
  
  dplyr::mutate(flank_distance = abs(as.integer(flank_distance))) %>%
  dplyr::group_by(peak) %>%
  dplyr::slice_min(flank_distance, n = 3)

sum(!is.na(anno.test.df.unnest$flank_genes))
anno.test.df.unnest[500,]
lookup
data.frame(lookup) %>% tidyr::pivot_longer()

lookup <- AnnotationDbi::intraIDMapper(anno.test.df.unnest$flank_genes, 'DROME', srcIDType = 'FLYBASE', destIDType = 'SYMBOL') 
 #closing.osaBound.entrez <- clusterProfiler::bitr(closing.osaBound$Geneid, fromType = 'FLYBASE', toType = 'ENTREZID', OrgDb = org.Dm.eg.db)
#fairePeaks.dm6 <- fairePeaks %>%
#   purrr::pmap_dfr(function(...) {
#     current <- tibble::tibble(...)
#     distances <- GenomicRanges::distance(GRanges(current), dm6.genes) 
#     df <- data.frame(Geneid = dm6.genes$gene_id, distances = distances)
#     nearbyGenes <- dplyr::slice_min(df, distances, n = 4)
#     out <- current %>% 
#       dplyr::mutate(nearbyGenes = list(nearbyGenes)) 
#     
#     return(out)
#   }) %>%
#  tidyr::unnest(nearbyGenes) %>%
#  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>% # join the rna seq data by Geneid
#  # average transcription at 3LW and 24h APF then categorize gene by 'up', 'down', or 'static'
#  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
#                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
#                l3v24 = log2(L3_mean / h24_mean),
#                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
#                                       l3v24 <= -1 ~ 'up',
#                                       T ~ 'static'))  


```


```{r}
closing.down.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & L3_mean >= 2 & cat == 'down' & osa.cnr & !yw.cnr) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 

closing.static.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & L3_mean >= 2 & cat == 'static' & osa.cnr & !yw.cnr) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 


static.osaUnBound <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'static') %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>% # not bound by osa specifically 
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>% # gene must be expressed at L3 or 24h
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'static.osaUnBound')

down.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(L3_mean >= 2 & cat == 'down' & osa.cnr & !yw.cnr) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'down.osaBound')


down <- fairePeaks.dm6 %>%
  dplyr::filter(L3_mean >= 2 & cat == 'down') %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 

###
closing.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr) %>% # closing category, bound specifically by osa
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>% # gene must be expressed at L3 or 24h
  dplyr::select(Geneid) %>%
  dplyr::distinct() 

closing <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>% # gene must be expressed at L3 or 24h
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'closing')

static.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'static' & osa.cnr & !yw.cnr) %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>% # gene must be expressed at L3 or 24h
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'static.osaBound')

static <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'static') %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>% # gene must be expressed at L3 or 24h
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'static')
####



static.osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(L3_mean >= 2 & faireCat.3LW_24h == 'static' & osa.cnr & !yw.cnr) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'static.osaBound')

static.down <- fairePeaks.dm6 %>%
  dplyr::filter(L3_mean >= 2 & faireCat.3LW_24h == 'static' & cat == 'down') %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'static.down.osaBound')

osaBound <- fairePeaks.dm6 %>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() 

osaBound

allFAIRE <- fairePeaks.dm6 %>%
#  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct() %>%
  dplyr::mutate(grp = 'osaUnBound')

closing.down <- fairePeaks.dm6 %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & L3_mean >= 2 & cat == 'down') %>%
  dplyr::select(Geneid) %>%
  dplyr::distinct()
```

```{r}
closing.osaBound.PLOT <- goEnrich(closing.osaBound, closing, 'closing.osaBound')
static.osaBound.PLOT <- goEnrich(static.osaBound, osaBound, 'static.osaBound')

ggsave(plot = closing.osaBound.PLOT, file = 'rPlots/go-enrich-closing_osaBound.png', width = 10, height = 10, dpi = 300)
ggsave(plot = static.osaBound.PLOT, file = 'rPlots/go-enrich-static_osaBound.png', width = 10, height = 10, dpi = 300)
```


```{r}
osaBound.PLOT <- goEnrich(osaBound, allFAIRE, 'osaBound')
osaBound.PLOT
ggsave(plot = osaBound.PLOT, file = 'rPlots/go-enrich-osaBound.png', width = 10, height = 8, dpi = 300)
```
```{r}

#osaBound.entrez <- select(org.Dm.eg.db, keys = osaBound$Geneid, columns = c('SYMBOL','GO','ENTREZID'), keytype = 'FLYBASE') %>% dplyr::distinct(ENTREZID)

closing.osaBound.semsim <- GOSemSim::mgeneSim(closing.osaBound$Geneid, semData = goData, verbose = F)

test <- closing.osaBound.semsim %>% data.frame() %>% tibble::rownames_to_column() %>% tidyr::pivot_longer(!rowname, names_to = 'colname')
test.symbol <- select(org.Dm.eg.db, keys = test$rowname, columns = 'SYMBOL', keytype = 'FLYBASE') %>% dplyr::distinct()

out <- test %>%
  dplyr::rowwise() %>%
  dplyr::mutate(rowSymbol = test.symbol[rowname == test.symbol$FLYBASE,]$SYMBOL,
                colSymbol = test.symbol[colname == test.symbol$FLYBASE,]$SYMBOL)

out %>%
#  dplyr::group_by(cat) %>%
 # dplyr::filter(cat == 'down') %>%
  tidyHeatmap::heatmap(rowSymbol, colSymbol, value, 
                       palette_value = c('white','blue','red'),
                       row_km = 6,
                       column_km = 6) %>% tidyHeatmap::save_pdf('test-down.pdf')
 
  

 closing.osaBound.entrez <- clusterProfiler::bitr(closing.osaBound$Geneid, fromType = 'FLYBASE', toType = 'ENTREZID', OrgDb = org.Dm.eg.db)
clusterProfiler::setReadable(clusterProfiler::enrichWP(closing.osaBound.entrez$ENTREZID, organism = 'Drosophila melanogaster'),
                             org.Dm.eg.db, keyType = 'ENTREZID')


```


##### GO enrichment summary - osaBound closing nearest downreg genes

-   changing to use perspective of faire peaks

```{r}

ggo %>%
  data.frame() %>%
  dplyr::filter(ONTOLOGY == 'BP') %>%
  dplyr::mutate(SYMBOL = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(SYMBOL) %>%
  dplyr::mutate(simple.description = dplyr::case_when(grepl('dendrite|neur|nerv|sensory',Description) ~ 'Nervous System',
                                                      grepl('wing|disc', Description) ~ 'Imaginal Disc Development',
                                                      grepl('metamor|morpho',Description) ~ 'Metamorphosis/Morphogenesis',
                                                      grepl('devel|multicell|fate', Description) ~ 'Development',
                                                      T ~ Description)) %>%
  dplyr::group_by(simple.description) %>%
  dplyr::mutate(avg.p = mean(p.adjust),
                newcount = dplyr::n_distinct(SYMBOL)) %>%
  dplyr::ungroup() %>%
  dplyr::mutate(simple.description = factor(simple.description),
                simple.description = forcats::fct_reorder(simple.description, avg.p, .desc = T),
                Description = factor(Description),
                Description = forcats::fct_reorder(Description, p.adjust, .desc = T),
                logp = log10(p.adjust)) %>%
  dplyr::filter(p.adjust <= 0.015) %>%
  #ggplot(aes(logp, Description, color = simple.description )) +
  ggplot(aes(logp, Description, )) +
  geom_point(size = 6) +
  #scale_color_manual(values = cbPalette.4) +
  theme(axis.text = element_text(size = 12),
        axis.title.y = element_blank())
ggsave('rPlots/osabound-closing-nearDownGene-GO-summary.png', width = 8, height = 6, dpi = 300)
```



### e93 sensitive sites

doesn't look like there is a correlation between osa binding and sites
that are sensitive to e93 in discs using liftover to convert dm3
coordinates from geo data, so may be worth verifying with my own data...
would have to align and process e93 gof chip and faire data

```{r}
e93_faire <- read.csv('~/McKay/FAIRE_E93/GEO_data/GSE141738_union_chip_peaks_annotated.csv')

e93_faire_sensitive <- e93_faire %>%
  dplyr::filter(e93_sensitive == 'sensitive') %>%
  #dplyr::filter(e93_sensitive_behavior == 'Decreasing') %>%
  dplyr::select(chr, start, end)

write.table(e93_faire_sensitive, 'e93-faire-sensitive_dm3.bed', col.names = F, row.names = F, quote = F, sep = '\t')

e93_faire_sensitive_closing.dm6 <- read.csv('~/McKay/BAP_Project/R/e93-faire-sensitive-closing_dm6.bed', sep = '\t', header = F) %>%
  setNames(c('seqnames','start','end','other'))

e93_faire_sensitive.dm6 <- read.csv('~/McKay/BAP_Project/R/e93-faire-sensitive_dm6.bed', sep = '\t', header = F) %>%
  setNames(c('seqnames','start','end'))


peaks %>% 
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse' & faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr) %>%
  #dplyr::group_by(peak) %>%
  dplyr::mutate(e93_sensitive_decreasing = ifelse(GRanges(.) %over% GRanges(e93_faire_sensitive_closing.dm6), T, F))

ifelse(GRanges(e93_faire_sensitive_closing.dm6) %over% GRanges(peaks %>% 
  dplyr::filter(assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse' & faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr)), T, F) %>% 
  data.frame() %>% 
  setNames('test') %>%
  dplyr::group_by(test) %>%
  dplyr::tally()

ifelse(GRanges(e93_faire_sensitive.dm6) %over% osaPeaks, T, F) %>% 
  data.frame() %>% 
  setNames('test') %>%
  dplyr::group_by(test) %>%
  dplyr::tally()

peaks
ifelse(GRanges(peaks) %over% GRanges(e93_faire_sensitive_closing),T,F)
```

### fraction of osa binding sites by faire

while most sites that close between 3LW and 24h are bound by Osa at 3LW
most Osa binding sites are static from 3LW to 24h, only \~30% of osa
sites close.

```{r}
osaPeaks %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(frac = n/sum(n))
```

### osa bound closing nearest gene expression dynamics

Are the osa bound sites that close late near genes that are temporally
dynamic and in what direction?

Most of the genes near osa binding sites that close are static between
3LW and 24h, however these closing sites are more frequently near genes
that go down in expression than sites that open or are static.

```{r}
#TODO - use deseq to better identify differential genes -- method below may not be accurate
##### 
#wing.rnaseq %>%
#  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
#                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
#                l3v24 = log2(L3_mean / h24_mean),
#                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
#                                       l3v24 <= -1 ~ 'up',
#                                       T ~ 'static')) %>%
#  dplyr::filter(L3_mean >= 2 | h24_mean >= 2) %>%
#  dplyr::group_by(cat) %>%
#  dplyr::tally()

closing.osaBound.rna <- osaPeaks %>%
  data.frame() %>%
#  dplyr::filter(faireCat.3LW_24h == 'closing' & grepl('Promoter',anno.new)) %>%
  data.table::setnames('geneId','Geneid') %>%
  dplyr::select(Geneid, SYMBOL, faireCat.3LW_24h, anno.new) %>%
#  dplyr::distinct() %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>%
  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
                l3v24 = log2(L3_mean / h24_mean),
                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
                                       l3v24 <= -1 ~ 'up',
                                       T ~ 'static')) 

closing.osaBound.rna %>%
  dplyr::filter(L3_mean >= 2 | h24_mean >= 2,
                grepl('Promoter|Intron|Exon|Distal',anno.new),
                faireCat.3LW_24h != 'NA') %>%
  #dplyr::filter(l3v24 >= 1) %>%
  #dplyr::group_by(SYMBOL) %>%
  #dplyr::tally() %>%
  #dplyr::arrange(desc(n))
  dplyr::group_by(faireCat.3LW_24h, cat) %>%
  dplyr::tally() %>%
  dplyr::mutate(frac = n/sum(n)) %>%
  ggplot(aes(faireCat.3LW_24h, frac, fill = cat, label = n)) +
  geom_bar(stat = 'identity') +
  geom_text(position = position_stack(vjust = 0.5), size = 10) +
  scale_fill_manual(values = c('#e8335d','grey80','#5ae342'), name = 'DEG 3LW-24h') +
  theme(axis.title = element_blank(),
        axis.text = element_text(size = 20))
  
#wing.rnaseq %>%
#  dplyr::filter(Geneid == 'FBgn0260642') %>%
#  tidyr::pivot_longer(!Geneid) %>%
#  dplyr::group_by(name) %>%
#  dplyr::mutate(stage = stringr::str_remove(name, 'X'),
#                stage = stringr::str_split_fixed(stage, '_', n = 2)[[1]],
#                stage = factor(stage, levels = c('L3','6h','18h','24h','36h','44h'))) %>%
#  ggplot(aes(stage, value)) +
#  geom_point()
  
                
```

```{r}
#####

osaPeaks.closing.rna <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  data.table::setnames('geneId','Geneid') %>%
  dplyr::select(Geneid, SYMBOL) %>%
#  dplyr::distinct() %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>%
  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
                l3v24 = log2(L3_mean / h24_mean),
                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
                                       l3v24 <= -1 ~ 'up',
                                       T ~ 'static')) 
osaPeaks.closing.rna %>%
  dplyr::group_by(cat) %>%
  dplyr::tally() %>%
  dplyr::mutate(frac = n/sum(n))

#####

closing.osaBound.rna <- closing.osaBound %>%
  data.frame() %>%
  dplyr::rename('Geneid' = 'geneId') %>%
  dplyr::select(Geneid, SYMBOL) %>%
#  dplyr::distinct() %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>%
  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
                l3v24 = log2(L3_mean / h24_mean),
                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
                                       l3v24 <= -1 ~ 'up',
                                       T ~ 'static')) %>%
  dplyr::filter(mean  >= 2)

closing.osaBound.rna %>%
  dplyr::group_by(cat) %>%
  dplyr::tally()
  
osaBound.rna <- osaPeaks %>%
  data.frame() %>%
  dplyr::rename('Geneid' = 'geneId') %>%
  dplyr::select(Geneid, SYMBOL) %>%
#  dplyr::distinct() %>%
  dplyr::left_join(., wing.rnaseq, by = 'Geneid') %>%
  dplyr::mutate(L3_mean = rowMeans(.[,c('L3_rep1', 'L3_rep2')]),
                h24_mean = rowMeans(.[,c('X24h_rep1', 'X24h_rep2')]),
                l3v24 = log2(L3_mean / h24_mean),
                cat = dplyr::case_when(l3v24 >= 1 ~ 'down',
                                       l3v24 <= -1 ~ 'up',
                                       T ~ 'static')) %>%
  dplyr::filter(mean  >= 2)

osaBound.rna %>%
  dplyr::group_by(cat) %>%
  dplyr::tally() %>%
  dplyr::mutate(frac = n/sum(n))
  

osaPeaks.closing.downGene %>%
  dplyr::group_by(SYMBOL) %>%
  dplyr::tally() %>%
  dplyr::mutate(symbol = forcats::fct_reorder(SYMBOL, n)) %>%
  dplyr::filter(n>=3) %>%
  ggplot(aes(n, symbol)) +
  geom_point()

osaPeaks.closing.downGene %>%
  dplyr::group_by(SYMBOL) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
closing.go <- select(org.Dm.eg.db, 
       keys = osaPeaks.closing.downGene$Geneid, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

closing.terms <- select(GO.db, 
       keys = closing.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

closing <- dplyr::bind_cols(closing.go, closing.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

closing.entrez <- closing %>% .$ENTREZID

ggo.closing <- clusterProfiler::enrichGO(gene = closing.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.closing %>%
  data.frame() 
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol) %>%
  dplyr::select(symbol) %>%
  dplyr::distinct() %>%
  dplyr::arrange(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
#png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.closing, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
#dev.off()


  #dplyr::group_by(cat) %>%
  #dplyr::tally() %>%
  #ggplot(aes('', n, fill = cat)) + 
  #geom_bar(stat = 'identity', position = 'fill')
#
  #dplyr::group_by(cat,l3v24) %>%
  #dplyr::tally()
  #reshape2::melt() %>%
  #  dplyr::rowwise() 
  #  dplyr::mutate(variable = stringr::str_replace(variable, 'X', ''),
  #                rna_grp = stringr::str_split_fixed(variable, pattern = '_', n = 2)[1],
  #                rna_grp = factor(rna_grp, levels = c('L3','6h','18h','24h','36h','44h'))) %>%
  #dplyr::filter(!is.na(rna_grp)) %>%
  #  dplyr::group_by(cat) %>%
  #dplyr::tally() 
  #ggplot(aes(cat,))
#
  #  dplyr::summarise(mean = mean(value)) %>%
  #  dplyr::mutate(log_mean = log10(1 + mean),
  #                SYMBOL = forcats::fct_reorder(SYMBOL, log_mean)) %>%
  #dplyr::filter(cat == 'down') %>%
  #ggplot(aes(rna_grp, SYMBOL, fill = log_mean)) +
  #geom_raster()
  #geom_point(alpha = 0.5)
  #geom_freqpoly(bins = 10) 
#
  #theme(legend.position = 'none') 
  

```

### aef1 motif scan idea

```{r}

```

### br fimo

```{r}
#testing with peak summits by rep
# using summit coordinates doesn't seem to majorly impact results
#summit_files <- purrr::map_df(c('rep1','rep2'), function(x) {
#  fp <- paste0('Peaks/osaGFP-3LW-wing-aGFP-CnR-sup-',x,'_dm6_trim_q5_dupsRemoved_allFrags_summits.bed')  
#  read.table(fp) %>%
#    setNames(c('seqnames', 'start', 'end', 'path', 'score')) %>%
#    dplyr::mutate(grp = x) %>%
#    dplyr::filter(GRanges(.) %over% GRanges(peaks %>% dplyr::filter(assay == 'cnr' & osa.cnr & !yw.cnr)))
#}) %>% GRanges()

#streme
#summit.seq.byRep <- summit_files %>% 
#  split(mcols(.)$grp) %>%
#  resize(200, 'center') %>%
#  memes::get_sequence(dm6)
pan_meme <- meme_db_on %>% dplyr::filter(grepl('pan',altname))
pan_meme
brD.seq <- brD %>% memes::get_sequence(dm6)

brFimo <- memes::runFimo(brD.seq, universalmotif::to_list(pan_meme))
brFimo
brFimo %>% 
  data.frame() %>% 
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_', n =2)[[1]]) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
```

### motifs in osa peaks

#### osa bound 'enhancers'
Trying motif enrichment within the "osa bound enhancer" subgroup
so - osa peaks within 3lw faire peaks annotated as distal or intronic - ie. not promoters or exons
As a control, using any non osa bound 3lw faire peaks

-note that using osa bound faire peaks produces pan and lola as enriched motif
```{r}

osa.enhancer.seq <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  #dplyr::filter(WT.3LW) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.seq <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  #dplyr::filter(WT.3LW) %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.ame <- osa.enhancer.seq %>%
  memes::runAme(control = osa.seq) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

osaBound.ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:10) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 

ggsave('rPlots/3E_osaBound-enhancer.png', width = 3, height = 4, dpi = 300)
```
#### osa bound open 3lw enhancers
what motifs are enriched within osa bound "closed" 3lw sites annotated as distal or intronic - ie. "enhancers"
```{r}

osa.enhancer.seq <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(WT.3LW) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.seq <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(WT.3LW) %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.ame <- osa.enhancer.seq %>%
  memes::runAme(control = osa.seq) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

osaBound.ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:10) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 

ggsave('rPlots/3E_osaBound-enhancer-3lwPeak.png', width = 2.8, height = 2.5, dpi = 300)
```
#### osa bound closed 3lw enhancers
what motifs are enriched within osa bound "closed" 3lw sites annotated as distal or intronic - ie. "enhancers"
```{r}

osa.enhancer.seq <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(!WT.3LW) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.seq <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(!WT.3LW) %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.ame <- osa.enhancer.seq %>%
  memes::runAme(control = osa.seq) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

osaBound.ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:5) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 

ggsave('rPlots/3E_osaBound-enhancer-no3lwPeak.png', width = 2.5, height = 2.5, dpi = 300)
```
#### osa bound vs unbound AME 
ame enrichment of osa bound sites vs all potential regulatory sites not bound by osa 
(ie. FAIRE peaks across wing development)
```{r}
osa.enhancer.seq <- osaPeaks %>%
  data.frame() %>%
  #dplyr::filter(!WT.3LW & !WT.6h & !WT.24h & !WT.44h) %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h) %>%
  #dplyr::filter(WT.3LW | WT.6h | WT.24h | WT.44h) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.seq <- fairePeaks.wt %>%
  data.frame() %>%
  #dplyr::filter(WT.3LW & !WT.6h & !WT.24h & !WT.44h) %>%
  #dplyr::filter(!WT.3LW & !WT.6h & !WT.24h & !WT.44h) %>%
  #dplyr::filter(WT.3LW | WT.6h | WT.24h | WT.44h) %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.ame <- osa.enhancer.seq %>%
  memes::runAme(control = osa.seq) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

osaBound.ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:10) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 

ggsave('rPlots/3E_osaBound.png', width = 3, height = 4, dpi = 300)
```
#### osa bound closing vs static FAIRE "enhancer" peaks
```{r}
query <- fairePeaks.wt %>%
  data.frame() %>%
  #dplyr::filter(faireCat.3LW_24h == 'closing' & anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter( osa.cnr & !yw.cnr) %>%
  dplyr::filter(WT.3LW & !WT.6h & !WT.24h) %>%
  #dplyr::filter(WT.3LW | WT.6h | WT.24h | WT.44h) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

control <- fairePeaks.wt %>%
  data.frame() %>%
  #dplyr::filter(faireCat.3LW_24h == 'static') %>%
  #dplyr::filter(!WT.3LW) %>%
  dplyr::filter(WT.3LW & WT.24h) %>%
  #dplyr::filter(anno.new %in% c('Distal Intergenic', 'Intron')) %>%
  #dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

ame <- query %>%
  memes::runAme(control = control) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:10) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 



#ggsave('rPlots/3E_osaBound.png', width = 3, height = 4, dpi = 300)
```

```{r}
#```{r}
#using a slightly larger range becuase average peak size is closer to ~300bp - is this the right call?
osa.enhancer.seq <- fairePeaks %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.seq <- fairePeaks %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & anno.new %in% c('Distal Intergenic','Intron')) %>%
  dplyr::filter(!osa.cnr | osa.cnr & yw.cnr) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)


osaBound.ame <- osa.enhancer.seq %>%
  memes::runAme(control = osa.seq) %>%
  dplyr::group_by(motif_alt_id) %>%
  dplyr::mutate(symbol = stringr::str_split_fixed(motif_alt_id, '_|-', n = 2)[[1]]) 

osaBound.ame %>%
  dplyr::group_by(symbol) %>%
  dplyr::filter(adj.pvalue == min(adj.pvalue) & rank %in% 1:15) %>%
  dplyr::mutate(neglogP = -log10(adj.pvalue)) %>%
  dplyr::ungroup() %>% # fct_reorder fails with grouped data, so annoying
  dplyr::mutate(symbol = forcats::fct_reorder(symbol, neglogP)) %>%
  ggplot(aes('Osa Bound', symbol, fill = neglogP)) +
  geom_raster() +
  scale_fill_continuous(type = 'viridis', name = '-log(adj.Pvalue)') +
  theme(axis.text.y = element_text(size = 12),
        axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) 
ggsave('rPlots/3E_osaBound-faire-closing.png', width = 4, height = 4, dpi = 300)
 
```

```{r}
osaBound.dreme %>%
  dplyr::mutate(consensus = forcats::fct_reorder(consensus, eval)) %>%
  dplyr::filter(nsites >= 750) %>%
  ggplot(aes(eval, consensus, color = nsites)) + 
  geom_point(size = 5)


osa.tomtom <- memes::runTomTom(osaBound.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 

tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osaBound-dreme_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```

#### promoters control

```{r}
dm6.promoters <- promoters(dm6.TxDb) %>%
  data.frame() %>%
  dplyr::filter(seqnames %in% c('chr2L','chr2R','chr3L','chr3R','chrX')) %>%
  GRanges() %>%
  resize(2000, "center") 

dm6.promoters.seq <- dm6.promoters %>%
  memes::get_sequence(dm6) %>%
  data.frame() 

colnames(dm6.promoters.seq) <- 'seqs'

dm6.promoters.split <- stringr::str_split(dm6.promoters.seq$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(dm6.promoters.split) <- as.character(c(1:2000))

dm6.promoters.df <- dplyr::bind_cols(dm6.promoters.split, data.frame(dm6.promoters))

dm6.promoters.df

dm6.promoters.df %>%
 # dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
  tidyr::pivot_longer(c(1:2000)) %>%
  dplyr::mutate(name = dplyr::case_when(strand == '+' ~ as.integer(name)-1000,
                                        strand == '-' ~ -(as.integer(name)-1000))) %>%
  dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
                AT = ifelse(value == 'T' | value == 'A', T, F),
                name = as.integer(name)) %>%
  dplyr::group_by(name, AT) %>%
  dplyr::tally() %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  dplyr::filter(AT) %>%
  ggplot(aes(name, freq)) +
  ylim(c(0,1)) +
  geom_point(alpha = 0.2) +
  geom_smooth(method = 'loess', span = 0.1)
```

#### at string search

```{r}
#####

osaBound.center.seq = osaPeaks %>% 
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.flank.up.seq = osaPeaks %>%
  shift(250) %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osaBound.flank.down.seq = osaPeaks %>%
  shift(-250) %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

# freqIslands = vcountPattern("CG", seqChr8Islands) / width(seqChr8Islands) 
osaBound.center.at <- Biostrings::vcountPattern('AT', osaBound.center.seq) / width(osaBound.center.seq) %>%
  data.frame(values = .)

osaBound.center.at %<>%
  dplyr::mutate(grp = 'center')

osaBound.flank.up.at <- Biostrings::vcountPattern('AT', osaBound.flank.up.seq) / width(osaBound.flank.up.seq) %>%
  data.frame(values = .) 

osaBound.flank.up.at %<>%
  dplyr::mutate(grp = 'flank-up')

osaBound.flank.down.at <- Biostrings::vcountPattern('AT', osaBound.flank.down.seq) / width(osaBound.flank.down.seq) %>%
  data.frame(values = .) 

osaBound.flank.down.at %<>%
  dplyr::mutate(grp = 'flank-down')

test <- dplyr::bind_rows(osaBound.center.at,
                 osaBound.flank.up.at,
                 osaBound.flank.down.at)

test %>%
  ggplot(aes(grp, values)) +
  geom_violin() +
  geom_point(alpha = 0.4, position = position_jitter(width = 0.2))
```

#### other

```{r}


test.mat <- osaBound.seq.split %>%
#osaBound.seq.split %>%
#  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
#  dplyr::filter(!WT.3LW) %>%
  #dplyr::select(id, WT.3LW, c(60:4059)) %>%
#  dplyr::select(c(60:2058)) %>%
  purrr::map(., function(x) {
    dplyr::case_when( x == 'A' | x == 'T' ~ 1, x == 'G' | x == 'C' | x == 'N' ~ 0)
  }) %>%
  dplyr::bind_rows(.) 
  

test.mat %>%
  dplyr::mutate(peak = 1:nrow(.)) %>%
  tidyr::pivot_longer(!peak) %>%
#  dplyr::mutate(value = factor(value)) %>%
  dplyr::group_by(value, name) %>%
  dplyr::tally() %>%
  dplyr::filter(value == 1) %>%
  dplyr::mutate(name = as.integer(name),
                avg = n / 1953) %>%
 # dplyr::arrange(name)
  ggplot(aes(name,avg )) +
  geom_point()
  geom_tile() +
#  scale_fill_manual(values = c('white','red')) +
  theme(axis.text.x = element_blank())

#ComplexHeatmap::Heatmap(test.mat, cluster_rows = F )
#pheatmap::pheatmap(test.mat)
```

#### bedtools nuc control test

writes bed files for regions at center and flanking osa peaks -- need to
then run bedtools nuc on those bed files with dm6 fasta then plots the
average AT content within those windows. Just a sanity check that I can
get a similar result to what I have above.

```{r}
#source('~/NystLib/R/peakUtils.R')
#
##ctbp <- readNarrowPeak('~/McKay/CtBP_CnR/Peaks/yw_3LW_wing_aCtBP_sup-Rep1_dm6_trim_q5_dupsRemoved_allFrags_peaks.narrowPeak') %>%
#
#
#####
osaPeaks.bed <- osaPeaks %>%
  resize(250, 'center') 

osaPeaks %>% shift(500) %>%
  resize(250, 'center') %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_up250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')

osaPeaks %>% shift(-500) %>%
  resize(250, 'center') %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_down250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')
#seqlevelsStyle(osaPeaks.bed) <- 'NCBI'

osaPeaks.bed %>%
  data.frame() %>%
  dplyr::select(seqnames, start, end) %>%
  write.table(file = 'osaPeaks_250bp.bed', col.names = F, row.names = F, quote = F, sep = '\t')


center <-  read.csv('osaPeaks_250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'center')

up <-  read.csv('osaPeaks_up250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'up')

down <-  read.csv('osaPeaks_down250bp_nuc.tsv', sep = '\t') %>%
  dplyr::mutate(cat = 'down')

dplyr::bind_rows(up, center, down) %>%
  dplyr::select(X4_pct_at, cat) %>%
  dplyr::group_by(cat, X4_pct_at) %>%
  ggplot(aes('',X4_pct_at, fill = cat)) +
  geom_violin()
```

#### test control

```{r}
#k27me3 <- readNarrowPeak('~/McKay/CtBP_CnR/Peaks/yw_3LW_wing_aK27me3_sup-Rep1_dm6_trim_q5_dupsRemoved_allFrags_peaks.narrowPeak') %>%
k27me3 <- readNarrowPeak('~/McKay/ecrGFSTF-3LW-Wing-CnR/Peaks/GSE202807_EcRGFSTF_3LW_WING_CnR_antiFlag_Merge_Sup_dm6_trim_q5_dupsRemoved_20to120_peaks.narrowPeak') %>%
#ctbp <- readNarrowPeak('/Volumes/McKay/Pierre : Lab/snystrom/bioinformatics3001-2_workstation_backup/E93_GOF_Project/ChIP_Data/E93_GOF_ChIP/Peakfiles/E93_GOF_ChIP_2Reps_FAIRE_PooledPeaks_peaks.narrowPeak') %>%
  data.frame() %>%
  dplyr::filter(!seqnames %in% c('chr4','chrY','chrM')) %>%
  GRanges() %>%
  resize(width = 1000, fix = 'center') %>%
  memes::get_sequence(dm6)

k27me3.df <- k27me3 %>%
  data.frame() 

colnames(k27me3.df) <- 'seqs'


k27me3.split <- stringr::str_split(k27me3.df$seqs, pattern = '', simplify = T) %>%
  data.frame()

colnames(k27me3.split) <- as.character(c(1:1000))

#test <- dplyr::bind_cols(data.frame(osaPeaks), osaBound.seq.split)

k27me3.split %>%
#osaBound.seq.split %>%
  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
  #dplyr::filter(anno.new == "3' UTR") %>%
  #dplyr::select(anno.new, c(59:4058)) %>%
  tidyr::pivot_longer(!id) %>%
  dplyr::mutate(GC = ifelse(value == 'G' | value == 'C', T, F),
                AT = ifelse(value == 'T' | value == 'A', T, F),
                name = as.integer(name)) %>%
  dplyr::group_by(name, AT ) %>%
  dplyr::tally() %>%
  dplyr::mutate(freq = n/sum(n)) %>%
  dplyr::filter(AT) %>%
  ggplot(aes(name, freq)) +
  geom_point(alpha = 0.4)
#ggsave('rPlots/AT-K27me3_sup.png')
```

```{r}
#
####
#osaUnbound.seq.df <- osaUnbound.seq %>%
#  data.frame() 
#
#colnames(osaUnbound.seq.df) <- 'seqs'
#
#test <- stringr::str_split(osaUnbound.seq.df$seqs, pattern = '', simplify = T) %>%
#  data.table::data.table()
#
#colnames(test) <- as.character(c(1:50))
#
#test %>%
#  dplyr::mutate(id = 1:nrow(.), .before = 1) %>%
#  tidyr::pivot_longer(!id) %>%
#  dplyr::group_by(name, value) %>%
#  dplyr::tally() %>%
#  dplyr::mutate(name = as.integer(name),
#                value = factor(value, levels = c('A','T','G','C'))) %>%
#  ggplot(aes(name, n, fill = value)) +
#  geom_bar(stat = 'identity', position = 'fill')
#ggsave('rPlots/TEMP-osaUnbound-ATcontent.png')
#
####
osaBound.seq = osaPeaks %>% 
  resize(250, "center") %>%
  memes::get_sequence(dm6)

osa.streme.250 <- memes::runStreme(osaBound.seq, control = 'shuffle') 

osa.tomtom <- memes::runTomTom(osa.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 
osa.tomtom
osa.streme$motif
#####
tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osaBound_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```

### motif enrichment in osa dependent regions

I'm not convinced there is ver meaningful information in motif analysis
for osa cnr....

```{r}

###

#memes::
###


osa_closing.seq = fairePeaks.wt %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr & assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)

control_closing.seq = fairePeaks.wt %>% 
  data.frame() %>% 
  dplyr::filter(faireCat.3LW_24h == 'closing' & !osa.cnr & assay == 'faire' & experiment == 'WT FAIRE Wing Timecourse') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
 
osa.streme <- memes::runStreme(osa_closing.seq, control = control_closing.seq) 

###
osa.streme %>%
  dplyr::mutate(neg_log_evalue = -as.double(log_evalue),
                consensus = forcats::fct_reorder(consensus, neg_log_evalue, .desc = F)) %>%
  ggplot(aes(x=0,consensus, fill = neg_log_evalue)) +
  geom_tile(color = 'black') +
  scale_fill_viridis_c(guide = guide_legend(direction = 'horizontal', 
                                            nrow = 1, 
                                            label.position = 'bottom', 
                                            title.position = 'top')) +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 12),
        legend.position = 'bottom',
        legend.key.size = unit(0.2,'cm'))
ggsave('rPlots/motif-analysis/3-motif-consensus.png', width = 1.2, height = 2, dpi = 300, scale = 3)
###

#plot of enriched consensus significance
osa.tomtom <- memes::runTomTom(osa.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 


#####
tom.results <- tomResults(osa.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/osa-bound-closing_',unique(x$consensus),'.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```

### go analysis draft

```{r}
test <- osaPeaks %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing')
#  dplyr::filter(faire_osaDeGrad.log2FoldChange <= -1) %>%
#  dplyr::arrange(faire_osaDeGrad.log2FoldChange) 
  
 
dependent.go <- select(org.Dm.eg.db, 
       keys = test$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.dependent %>%
  data.frame() %>%
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
dev.off()

```

### osa cnr peak annotation

```{r, 3draft-1.png}
 peaks %>%
  dplyr::filter(assay == 'cnr' &  osa.cnr & !yw.cnr & faireCat.3LW_24h != 'NA') %>%
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
  ggplot(aes(x = faireCat.3LW_24h, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.01, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =3,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1,
                  ) +
  scale_fill_manual(values = cbPalette) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.3,"cm"),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/drafts/3draft-1.png', width = 4, height = 4, dpi = 300)

  
```

### peak annotation split by osa binding and dynamics

```{r, 3draft-2.png}

peaks %>%
  dplyr::filter(assay == 'faire' & faireCat.3LW_24h != 'NA' & experiment == 'WT FAIRE Wing Timecourse') %>%
  dplyr::mutate(osa.specific = ifelse(osa.cnr & !yw.cnr, T, F)) %>%
 # dplyr::filter(grepl('Distal|Exon|Intron|Promoter', anno.new)) %>%
 # dplyr::mutate(anno.temp = dplyr::case_when(grepl('Distal|Intron', anno.new) ~ 'Distal or Intronic',
 #                                           grepl('Exon', anno.new) ~ 'Exon',
 #                                           grepl('Promoter', anno.new) ~ 'Promoter',
 #                                           T~anno.new)) %>%
  dplyr::group_by(osa.specific, anno.new, faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n),
                pct = round(pct, digits = 2)) %>%
  ggplot(aes(x = n, y = anno.new, fill = osa.specific)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text(aes(label = n), position = position_fill(vjust = 0.5), color = 'black') +
  scale_fill_manual(values = c('#D53B59','#52CA3E','#1E82E0')) +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        axis.text.x = element_text(size = 5),
        legend.key.size = unit(0.3,"cm")) +
  facet_wrap(~faireCat.3LW_24h)

ggsave('rPlots/drafts/3draft-2.png', width = 6, height = 3, dpi = 300)
```

### blah

```{r}

```

# Fig 4

There is very little change between control and osa deGrad conditions.
Figure 4 should demonstrate this and highlight if there are any subtle
changes, where they occur, and how they relate to osa binding.

```{bash engine.opts='-l'}
multiBamSummary bins --bamfiles Bam/*deGrad*dupsRemoved.bam \
  -o deGrad-FAIRE-bamSummary.npz

plotCorrelation -in deGrad-FAIRE-bamSummary.npz \
  -c spearman -p heatmap --colorMap RdYlBu --plotNumbers \
  --labels control.1 deGrad.1 control.2 deGrad.2  \
  --removeOutliers --skipZeros \
  -o deGrad-FAIRE-spearman.png
```

## 4C - venn

```{r}
osaDeg <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(osaGFP.deGrad) %>%
  GRanges()

control <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::filter(osaGFP.control) %>%
  GRanges()

#ChIPpeakAnno::makeVennDiagram(list(osaDeg,control))
ChIPpeakAnno::makeVennDiagram(faire.osaDeg.byGrp)

####


```

## 4D

```{r}
### draft log2fold by degrad cat
#Might be useful to articulate that while there are some very subtle differences in osa degrad - there are very few twofold ectopic changes 
#ie. there is not an appreciable gain of accessibility genome-wide when osa is depleted

fairePeaks.deg %>%
  data.frame() %>%
  dplyr::mutate(twofold = ifelse(faire_osaDeGrad.log2FoldChange <= -1 | faire_osaDeGrad.log2FoldChange >= 1, T, F),
                cat = dplyr::case_when(osaGFP.control & !osaGFP.deGrad ~ 'control',
                                       !osaGFP.control & osaGFP.deGrad ~ 'osa degrad',
                                       T ~ 'other')) %>% 
  ggplot(aes(x = '',  faire_osaDeGrad.log2FoldChange, fill = faireCat.osaDeGrad)) + 
  geom_violin(position = 'identity', scale = 'count') +
    geom_point(aes(alpha = twofold), position = position_jitter(width = 0.1), color = 'red') +
  scale_alpha_manual(values = c(0,0.2)) +
  ylim(c(-2.25,2.25)) +
  scale_fill_manual(values = c('#C05066','#8AD04E','#608AE3')) +
  ylab("log2FoldChange(Control / OsaDeGrad)") +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_text(size = 12)) 

ggsave('rPlots/4-draft-5.png', width = 2, height = 4, dpi = 300)

```

## 4D

```{r}

fairePeaks.deg %>%
  data.frame() %>%
  dplyr::mutate(twofold = ifelse(faire_osaDeGrad.log2FoldChange <= -1 | faire_osaDeGrad.log2FoldChange >= 1, T, F),
                cat = dplyr::case_when(osaGFP.control & !osaGFP.deGrad ~ 'control',
                                       !osaGFP.control & osaGFP.deGrad ~ 'osa degrad',
                                       T ~ 'other')) %>% 
  ggplot(aes(x = '',  faire_osaDeGrad.log2FoldChange, fill = cat)) + 
  geom_violin(position = 'identity', alpha = 0.3, scale = 'count') +
  scale_alpha_manual(values = c(0,0.2)) +
  ylim(c(-2.25,2.25)) +
  scale_fill_manual(values = c('#C05066','#8AD04E','#608AE3')) +
  ylab("log2FoldChange(Control / OsaDeGrad)") +
  theme(legend.position = "none",
        axis.title.x = element_blank(),
        axis.text = element_text(size = 12)) 

ggsave('rPlots/4-draft-6.png', width = 2, height = 4, dpi = 300)

```

## 4E. FAIRE control vs osaDeGrad - split by degrad categories

Using deseq differential scoring between control and osaGFP deGrad to
define FAIRE peak categories. Using a very sensitive cutoff for log2fold
change. "ectopic" (gain of signal in degrad) \<= log2(control/degrad)
\<= -0.4, "dependent" (loss of signal in degrad) \>=
log2(control/degrad) \>= 0.4, "independet" (no significant change).

```{r, 4D}

#TODO - are heatmaps misleading? is this difficult to parse when split by these different peak call groups?


# union osa degrad peaks
deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- fairePeaks.deg %>%
  data.frame() %>%
  dplyr::group_by(faireCat.osaDeGrad) %>% #peak category assigned in peaks.R
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')

groups <- fairePeaks.deg$faireCat.osaDeGrad

lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4B.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'bottom', )
dev.off()

fairePeaks.deg %>% 
  data.frame() %>%
  dplyr::mutate(osa.cnr.specific = ifelse(osa.cnr & !yw.cnr,T,F)) %>%
  dplyr::group_by(faireCat.osaDeGrad, osa.cnr) %>%
  dplyr::tally() 
  dplyr::mutate(fraction = n / sum(n),
                change = faireCat.osaDeGrad) %>%
  dplyr::mutate(per = paste0(round(100*fraction, digits = 1),'%')) %>%
  plot.frac()


```

### no groups
```{r, 4D}



# union osa degrad peaks
deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)






deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,7))),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,7), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4B-nogroups.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()


```
### osa bound fraction
```{r}
fairePeaks.deg 
```

```{r}
fairePeaks.deg
# union osa degrad peaks
fairePeaks.deg.osaBound <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(osa.cnr & !yw.cnr) %>%
  dplyr::filter(faireCat.3LW_24h == 'closing') %>%
  GRanges() 

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.deg.osaBound, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- fairePeaks.deg.osaBound %>%
  data.frame() %>%
  dplyr::group_by(faireCat.osaDeGrad) %>% #peak category assigned in peaks.R
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')

groups <- fairePeaks.deg.osaBound$faireCat.osaDeGrad

lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(#textbox = anno_textbox(align_to = groups,
                                                    #                         as.list(deg.count),
                                                    #                         side = 'left',
                                                    #                         background_gp = gpar(fill = NA, col = NA),
                                                    #                         gp = gpar(col = 'black'), 
                                                    #                         padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1000, 0, 1000),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

#png('rPlots/4B.png', width = 5, height = 5, units = 'in', res = 300)
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups,  merge_legend = T, heatmap_legend_side = 'bottom', )
#dev.off()
```
```{r}
fairePeaks.deg %>%
  data.frame() %>%
  dplyr::mutate(neglogp = -log(faire_osaDeGrad.padj)) %>%
  #dplyr::filter(osa.cnr & !yw.cnr) %>%
  ggplot(aes(neglogp,faire_osaDeGrad.log2FoldChange, color = neglogp)) + 
#  geom_violin(draw_quantiles = T) 
  geom_point(position = position_jitter(width = 0.01)) 
#    stat_summary(fun = "mean", geom = "crossbar", width = 0.5,
#               colour = "red")
  
```

## 4F. osa dependent browser shot example

```{r, 4D}
#TODO - add highlights for dependent regions - currently manually annotating in illustrator 

#peaks %>%
#  dplyr::filter(assay == 'faire' & experiment == 'osaGFP deGrad Pupal Wing FAIRE' & faireCat.osaDeGrad == 'Osa Dependent') %>%
#  dplyr::filter(osaGFP.deGrad | osaGFP.control) %>%
#  dplyr::arrange(desc(faire_osaDeGrad.log2FoldChange)) %>%
#  ggplot(aes(faire_osaDeGrad.pvalue, faire_osaDeGrad.log2FoldChange)) +
#  geom_point()
#  dplyr::select(seqnames, start, end) %>%
#  write.table(., file = '~/Desktop/osaDep.bed', col.names = F, row.names = F, quote = F)
  
# chr2R:11,465,947-11,499,493 
inv <- data.frame(seqnames = 'chr2R',
                   start = 11465947,
                   end = 11499493) %>%
  GRanges()
  
#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("#50c42d","#c42d85","grey70","grey60","grey50")

deg.ss.browser <- faire.osaDeGrad.ssPool %>%
  dplyr::bind_rows(., faire.wt.ssPool[faire.wt.ssPool$grp == 'wt.3LW' | faire.wt.ssPool$grp == 'wt.24h' | faire.wt.ssPool$grp == 'wt.44h',]) %>%
  dplyr::mutate(color = track.colors)

deg.tracks <- get_cnr_tracks(deg.ss.browser, ylim = c(0,10))
deg.tracks <- c(deg.tracks$wt_3LW, 
                deg.tracks$wt_24h, 
                deg.tracks$wt_44h,
                deg.tracks$osaGFP.deGrad_control_faire, 
                deg.tracks$osaGFP.deGrad_nubG4_faire)

#brdisc.at <- Gviz::AnnotationTrack(range = inv, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 10000)

png('rPlots/4E-browser.png', width = 10, height = 5, res = 300, units = 'in')
Gviz::plotTracks(c(axis, deg.tracks, genetrack), chromosome = seqnames(inv) %>% as.character(), 
                 from = start(inv),
                 to = end(inv),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
  
```

## 4H. motif enrichment in osa dependent regions

```{r}

###

#memes::
###


osaDep.seq = fairePeaks.deg %>% 
  data.frame() %>% 
  #dplyr::filter(faireCat.3LW_24h == 'opening' & faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  #dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  #dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' & faire_osaDeGrad.log2FoldChange >= 1) %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
  
osaIndep.seq = fairePeaks.deg %>% 
  data.frame() %>% 
  #dplyr::filter(faireCat.3LW_24h == 'opening' & faireCat.osaDeGrad == 'Osa Independent') %>%
  dplyr::filter(faireCat.osaDeGrad == 'Osa Independent') %>%
  GRanges() %>%
  resize(250, "center") %>%
  memes::get_sequence(dm6)
 
osaDep.streme <- memes::runStreme(osaDep.seq, control = osaIndep.seq) 
#  dplyr::slice_min(pval, n = 8, with_ties = F)

#saDep.streme %<>%
#  dplyr::slice_min(evalue, n = 5, with_ties = F)
osaDep.streme %>%
  data.frame() %>%
  dplyr::mutate(consensus = forcats::fct_reorder(consensus, log_evalue)) %>%
  ggplot(aes(log_evalue, consensus)) + 
  geom_point()


  
###
osaDep.streme %>%
  dplyr::mutate(neg_log_evalue = -as.double(log_evalue),
                consensus = forcats::fct_reorder(consensus, neg_log_evalue, .desc = F)) %>%
  ggplot(aes(x=0,consensus, fill = neg_log_evalue)) +
  geom_tile(color = 'black') +
  scale_fill_viridis_c(guide = guide_legend(direction = 'horizontal', 
                                            nrow = 1, 
                                            label.position = 'bottom', 
                                            title.position = 'top')) +
  theme(axis.text.x = element_blank(), 
        axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(size = 12),
        legend.position = 'bottom',
        legend.key.size = unit(0.2,'cm'))
ggsave('rPlots/motif-analysis/4H-consensus.png', width = 1.2, height = 2, dpi = 300, scale = 3)
###

#plot of enriched consensus significance
osaDep.tomtom <- memes::runTomTom(osaDep.streme, database = '/Users/m/McKay/Motif_databases/FLY/fly_factor_survey.meme') %>%
  data.frame() 


#####
tom.results <- tomResults(osaDep.tomtom)

purrr::map(tom.results, function(x) {
  tom.df <- x %>%
    dplyr::group_by(neg_log_eval) %>%
    dplyr::arrange(neg_log_eval, .by_group = T) 
  
  ## plot the hm of tomtom results for current consensus 
  tom.hm <- tom.df %>%
    ggplot(aes(consensus, match_altname, fill = neg_log_eval)) +
    geom_tile() +
    scale_fill_continuous(type = 'viridis', name = '-log_eval', limit = c(-1.2,0)) +
    scale_y_discrete(expand = c(0,0)) +
    theme(panel.background = element_blank(),
          panel.grid = element_blank(), 
          panel.border = element_blank(),
          axis.title = element_blank(), 
          #axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1),
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm"), 
          legend.position = 'bottom') 
 
  ## format the tom.results for the rna heatmap
  rna.df <- format_rna(tom.df)
  
  ## plot the rna heatmap for current consensus 
  rna.hm <- rna.df %>%
    ggplot(aes(rna_grp,1, fill = log_mean)) +
    geom_tile() + 
    scale_fill_gradient(low = '#e9edf5', high = 'Red', limit = c(0,3)) +
    facet_grid(rows = vars(forcats::fct_rev(match_altname))) +
    theme(legend.position = "bottom",
          axis.text.y = element_blank(),
          plot.background = element_blank(),
          panel.background = element_blank(),
          panel.grid = element_blank(),
          axis.ticks = element_blank(),
          axis.title = element_blank(),
          strip.background = element_blank(),
          strip.text.y = element_blank(), 
          legend.key.size = unit(0.4, 'in'),
          plot.margin = margin(0,0,0,0, "cm")) 
  
  plot <- tom.hm + plot_spacer() + rna.hm + plot_layout(widths = c(4, -0.5 ,4.5))
  
  fp <- paste0('rPlots/motif-analysis/',unique(x$consensus),'_osaDependent.png')
  ggsave(fp,plot = plot, width = 5, height = 5, dpi = 300, unit = 'in')
  })
```

## 4I. Osa deGrad FAIRE signal in osa-bound l3-24h closing sites.

This plot shows that within the osa CnR peaks that close from 3LW to
24h, there is little to no change in their \~24h-36h accessibility when
osa is depleted.

```{r}

osaClosing <- fairePeaks.wt %>%
  data.frame() %>%
  dplyr::filter(faireCat.3LW_24h == 'closing' & osa.cnr & !yw.cnr) %>%
  GRanges()

deg.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., osaClosing, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(deg.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(deg.mats))
common_max = max(unlist(deg.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

deg.count <- osaClosing %>%
  data.frame() %>%
  dplyr::group_by(faireCat.osaDeGrad) %>%
  dplyr::tally() %>%
  .$n

names(deg.count) <- c('osa dependent', 'osa ectopic', 'osa independent')

groups <- osaClosing$faireCat.osaDeGrad

lgd = Legend(at = c("osa dependent", "osa ectopic", "osa independent"), legend_gp = gpar(fill = 2:4))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,1))),
                    left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(deg.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                    groups = anno_block(gp = gpar(fill = 2:4), show_name = F)))
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(ylim = c(0,1), axis = F)))
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

#png('rPlots/4C.png', width = 3, height = 3, units = 'in', res = 300)
ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
#dev.off()
```

## 4J. Browser shot of osa deGrad FAIRE at brDisc.

Shows that there is no increase in accessibility.

```{r, 4D}
#TODO - add 36h FAIRE data? better comparison for osa deGrad?

#track.colors <- RColorBrewer::brewer.pal(n = 4, "PiYG")
track.colors <- c("#50c42d","#c42d85","grey70","grey60","grey50")

deg.ss.browser <- faire.osaDeGrad.ssPool %>%
  dplyr::bind_rows(., faire.wt.ssPool[faire.wt.ssPool$grp == 'wt.3LW' | faire.wt.ssPool$grp == 'wt.24h' | faire.wt.ssPool$grp == 'wt.44h',]) %>%
  dplyr::mutate(color = track.colors)

deg.tracks <- get_cnr_tracks(deg.ss.browser, ylim = c(0,15))
deg.tracks <- c(deg.tracks$wt_3LW, 
                deg.tracks$wt_24h, 
                deg.tracks$wt_44h, 
                deg.tracks$osaGFP.deGrad_control_faire, 
                deg.tracks$osaGFP.deGrad_nubG4_faire)

brdisc.at <- Gviz::AnnotationTrack(range = brD, genome = 'dm6', name = 'brDisc', fill = 'grey', col = 'transparent', background.title = 'white', showTitle = F )

axis <- Gviz::GenomeAxisTrack(scale = 1000)

png('rPlots/4D.png', width = 4, height = 6, res = 300, units = 'in')
Gviz::plotTracks(c(axis, deg.tracks, brdisc.at), chromosome = seqnames(brD) %>% as.character(), 
                 from = start(brD + 1000),
                 to = end(brD + 1000),
                 cex.axis = 0.3,
                 cex.title = 0.5,
                 title.width = 1,
                 innerMargin = 10)
dev.off()
```

## Supp.

Using any peak call from the degrad experiment. So 'independent' regions
are accessible in both experiments.

```{r}

temp <- fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::filter(faire_osaDeGrad.log2FoldChange >= 0.9) %>%
  #dplyr::filter(faire_osaDeGrad.log2FoldChange >= 0.4 & faire_osaDeGrad.log2FoldChange < 0.9) %>%
  GRanges()
#fairePeaks.osaDependent <- fairePeaks.deg %>% data.frame() %>% dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent')

       #keys = fairePeaks.osaDependent$geneId, 
dependent.go <- select(org.Dm.eg.db, 
      # keys = fairePeaks.osaDependent$geneId, 
       keys = temp$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))


dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
unique(dependent$SYMBOL)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)

#png('rPlots/temp.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(ONTOLOGY~., scale = 'free')
#dev.off()



```

## 4E.

```{r}

osaDependent.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.osaDependent, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(osaDependent.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(osaDependent.mats))
common_max = max(unlist(osaDependent.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

#deg.count <- fairePeaks.deg %>%
#  data.frame() %>%
#  dplyr::group_by(faireCat.osaDeGrad) %>%
#  dplyr::tally() %>%
#  .$n
#
#names(deg.count) <- c('Osa Dependent', 'Osa Ectopic', 'Osa Independent')
#

#TODO - maybe cleaner to have this peak cat assigned globally in setup
groups <- fairePeaks.osaDependent %>% 
  data.frame() %>%
  dplyr::mutate(new.cat <- ifelse(osa.cnr & !yw.cnr, 'Osa Bound', 'Osa Unbound')) %>%
  .$new.cat

lgd = Legend(at = c("Osa Bound", "Osa Unbound"), legend_gp = gpar(fill = 2:3))

deg.hms <- purrr::map(names(osaDependent.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(osaDependent.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(osaDependent.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(osaDependent.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4E.png', width = 4, height = 3, units = 'in', res = 300)
#ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```

## 4F.

```{r}

osaEctopic.mats <- purrr::map(deg.bws, ~{
  normalizeToMatrix(., fairePeaks.osaEctopic, value_column = 'score', keep = c(0.11,0.99), extend = 1000 )
}) 

names(osaEctopic.mats) <- faire.ss %>% 
  dplyr::filter(experiment == 'osaGFP deGrad FAIRE') %>%
  dplyr::mutate(nm = paste0(stringr::str_split_fixed(id,'_|\\.', n = 5)[,3],
                            '.',
                            stringr::str_split_fixed(id,'_|\\.', n = 5)[,5])) %>%
  .$nm 

common_min = min(unlist(osaEctopic.mats))
common_max = max(unlist(osaEctopic.mats))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

groups <- fairePeaks.osaEctopic %>% 
  data.frame() %>%
  dplyr::mutate(new.cat <- ifelse(osa.cnr & !yw.cnr, 'Osa Bound', 'Osa Unbound')) %>%
  .$new.cat

lgd = Legend(at = c("Osa Bound", "Osa Unbound"), legend_gp = gpar(fill = 2:3))

deg.hms <- purrr::map(names(deg.mats), function(x) {
  if(grepl('control.Rep1',x)){
     EnrichedHeatmap(osaEctopic.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    name = 'zScore',
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), axis_param = list(side = 'left', facing = 'outside'), ylim = c(0,8))),
                    left_annotation = rowAnnotation(groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                    use_raster = F)
  }else{
    EnrichedHeatmap(deg.mats[[x]],
                    axis_name = c(-1, 0, 1),
                    pos_line = F,
                    column_title = x,
                    col = col_fun,
                    show_heatmap_legend = F, 
                    row_title = NULL,
                    top_annotation = HeatmapAnnotation(enriched = anno_enriched(gp = gpar(col = 2:4, lwd = 2), ylim = c(0,8), axis = F)), 
                    use_raster = F)
  }
})

names(deg.hms) <- names(deg.mats)

hms <- deg.hms$control.Rep1 + deg.hms$control.Rep2 + deg.hms$nubG4.Rep1 + deg.hms$nubG4.Rep2

png('rPlots/4F.png', width = 4, height = 3, units = 'in', res = 300)
#ComplexHeatmap::draw(hms, merge_legend = T, heatmap_legend_side = 'right', )
ComplexHeatmap::draw(hms,annotation_legend_list = c(lgd), row_split = groups, merge_legend = T, heatmap_legend_side = 'right', )
dev.off()
```

## 4H.

```{r}
#TODO - simplifiy - purrr::map
faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaDependent, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')
#
groups <- fairePeaks.osaDependent$faireCat.3LW_24h

#
lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                     left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.6h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '6h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) + 
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 

png('rPlots/4H.png', width = 4, height = 8, units = 'in', res = 300)
ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
dev.off() 
```

## 4I.

```{r}

faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaEctopic, value_column = 'score', keep = c(0.01,0.99), extend = 2000)

common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))

col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)

faire.count <- fairePeaks.osaEctopic %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h) %>%
  dplyr::tally() %>%
  dplyr::mutate(n = as.character(n)) %>%
  .$n

names(faire.count) <- c('closing','opening','static')
#
groups <- fairePeaks.osaEctopic$faireCat.3LW_24h

#
lgd = Legend(at = c("closing", "opening", "static"), legend_gp = gpar(fill = 2:4))
                             
hm <- EnrichedHeatmap(faire.3lw.mat, 
                      pos_line = F, 
                      axis_name = c(-2, 0, 2),
                      col = col_fun,
                      column_title = '3LW',
                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
                                                                               ylim = c(0,6),
                                                                               axis_param = list(side = 'left'))),  
                      name = 'FAIRE zScore',
                     left_annotation = rowAnnotation(textbox = anno_textbox(align_to = groups,
                                                                             as.list(faire.count),
                                                                             side = 'left',
                                                                             background_gp = gpar(fill = NA, col = NA),
                                                                             gp = gpar(col = 'black'), 
                                                                             padding = unit(0, 'mm')),
                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
                      row_title  = NULL,
                      use_raster = F) +
 EnrichedHeatmap(faire.6h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '6h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) + 
 EnrichedHeatmap(faire.24h.mat, 
                 pos_line = F, 
                 axis_name = c(-2, 0, 2),
                 col = col_fun,
                 column_title = '24h',
                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
                 row_title  = NULL,
                 show_heatmap_legend = F,
                 use_raster = F) 
 

png('rPlots/4I.png', width = 4, height = 8, units = 'in', res = 300)
ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
dev.off() 
```

```{r}
fairePeaks.osaDependent %>%
  data.frame() %>%
  dplyr::filter(!anno.new == "3' UTR") %>% #filtering this out because there are no peaks anno as 3'UTR in ectopic regins, messes up colors
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x= n, y = '', fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.08, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size = 4,
                  min.segment.length = 0,
                  force = 0.00,
                  force_pull = 1) +
  scale_fill_brewer(palette = 'Spectral') +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.background = element_blank(),) +
        #strip.text = element_blank())+
  facet_wrap(~faireCat.3LW_24h, ncol = 1)
ggsave('rPlots/4-draft-1.png', width = 5, height = 3, units = 'in', dpi = 300)
```

```{r}
fairePeaks.osaEctopic %>%
  data.frame() %>%
  dplyr::group_by(faireCat.3LW_24h, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x = n, y = '', fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.07, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =4,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  #scale_fill_manual(values = cbPalette) +
  scale_fill_brewer(palette = 'Spectral') +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.x = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank(),
        panel.background = element_blank(),
        strip.background = element_blank(),) +
        #strip.text = element_blank())+
  facet_wrap(~faireCat.3LW_24h, ncol = 1)
ggsave('rPlots/4-draft-2.png', width = 5, height = 3, units = 'in', dpi = 300)
```

## drafts
### deseq2 padj plot
```{r}
fairePeaks.deg %>%
  data.frame() %>%
  ggplot(aes(faire_osaDeGrad.log2FoldChange, faire_osaDeGrad.stat)) +
  geom_point() 
  geom_point(color = ifelse(fairePeaks.deg$faire_osaDeGrad.pvalue <= 0.05, 'red','black'))

# TODO - the adjusted p value plot is meaningless because almost all values ~ 1,
# if that's true then fine, but is this a 'bug' in pvalue correction with few-none differential regions?
fairePeaks.deg %>%
  data.frame() %>%
  ggplot(aes( log2FoldChange, pvalue)) +
  geom_point(color = ifelse(fairePeaks.deg$pvalue <= 0.05, 'red','black')) 
ggsave('rPlots/4-draft-log2Fold_vs_Pval.png', width = 4, height = 4, dpi = 300)

```

### peak annotations

```{r}
fairePeaks.deg %>%
    data.frame() %>%
#  dplyr::filter(faireCat.osaDeGrad == 'Osa Dependent' ) %>%
  dplyr::group_by(faireCat.osaDeGrad, anno.new) %>%
  dplyr::tally() %>%
  dplyr::mutate(pct = n/sum(n)) %>%
 # dplyr::filter(faireCat.osaDeGrad != 'Osa Ectopic') %>%
  ggplot(aes(x = faireCat.osaDeGrad, y = n, fill = anno.new)) +
  geom_bar(stat = 'identity', position = 'fill') +
  geom_text_repel(aes(label= ifelse(pct > 0.009, paste0(sprintf("%1.1f", pct*100),"%"), "")),
                     position=position_fill(vjust=0.5), colour="black", size =4,
                  min.segment.length = 0,
                  force = 0.01,
                  force_pull = 1) +
  scale_fill_brewer(palette = 'Spectral') +
  ylab('Fraction') +
  theme(axis.title.x = element_blank(),
        legend.title = element_blank(),
        legend.key.size = unit(0.5, 'cm'),
        axis.text.y = element_blank(),
        axis.ticks = element_blank(),
        axis.title.y = element_blank())
ggsave('rPlots/4G.png', width = 6, height = 5, units = 'in', dpi = 300)
```

```{r}
#fairePeaks.osaDep <- fairePeaks.deg %>% 
#  data.frame() %>%
#  dplyr::filter(faireCat.3LW_24h == 'opening') %>%
#  GRanges()
#
#faire.3lw.mat <- normalizeToMatrix(bws$wt_3LW, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#faire.6h.mat <- normalizeToMatrix(bws$wt_6h, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#faire.24h.mat <- normalizeToMatrix(bws$wt_24h, fairePeaks.osaDep, value_column = 'score', keep = c(0.01,0.99), extend = 2000)
#
#common_min = min(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
#common_max = max(c(faire.3lw.mat,faire.6h.mat,faire.24h.mat))
#
#col_fun = circlize::colorRamp2(seq(common_min, common_max, length.out = 6), viridis.hex)
#
##faire.count <- fairePeaks.osaDep %>%
##  data.frame() %>%
##  dplyr::group_by(faireCat.3LW_24h) %>%
##  dplyr::count() %>%
##  dplyr::mutate(n = as.character(n)) %>%
##  .$n
##
##names(faire.count) <- c('closing','opening','static')
##
#groups <- fairePeaks.osaDep$faireCat.osaDeGrad
#
##
#lgd = Legend(at = c("Osa Dependent", "Osa Ectopic", "Osa Independent"), legend_gp = gpar(fill = 2:4))
#                             
#hm <- EnrichedHeatmap(faire.3lw.mat, 
#                      pos_line = F, 
#                      axis_name = c(-2, 0, 2),
#                      col = col_fun,
#                      column_title = '3LW',
#                      top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4),
#                                                                               ylim = c(0,6),
#                                                                               axis_param = list(side = 'left'))),  
#                      name = 'FAIRE zScore',
#                     left_annotation = rowAnnotation(#textbox = anno_textbox(align_to = groups,
#                                                     #                        as.list(faire.count),
#                                                     #                        side = 'left',
#                                                     #                        background_gp = gpar(fill = NA, col = NA),
#                                                     #                        gp = gpar(col = 'black'), 
#                                                     #                        padding = unit(0, 'mm')),
#                                                      groups = anno_block(gp = gpar(fill = 2:4), show_name = F)),
#                      row_title  = NULL,
#                      use_raster = F) +
# EnrichedHeatmap(faire.6h.mat, 
#                 pos_line = F, 
#                 axis_name = c(-2, 0, 2),
#                 col = col_fun,
#                 column_title = '6h',
#                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
#                 row_title  = NULL,
#                 show_heatmap_legend = F,
#                 use_raster = F) + 
# EnrichedHeatmap(faire.24h.mat, 
#                 pos_line = F, 
#                 axis_name = c(-2, 0, 2),
#                 col = col_fun,
#                 column_title = '24h',
#                 top_annotation = HeatmapAnnotation(lines = anno_enriched(gp = gpar(col = 2:4), ylim = c(0,6))), 
#                 row_title  = NULL,
#                 show_heatmap_legend = F,
#                 use_raster = F) 
# 
#
##png('rPlots/4I.png', width = 4, height = 8, units = 'in', res = 300)
#ComplexHeatmap::draw(hm, merge_legend = T, annotation_legend_list = list(lgd), row_split = groups, heatmap_legend_side = 'bottom', )
##dev.off() 
```

### browser draft

### go analysis draft

```{r}
test <- fairePeaks.wt %>%
  data.frame() %>%
#  dplyr::filter(faire_osaDeGrad.log2FoldChange <= -1) %>%
  dplyr::arrange(faire_osaDeGrad.log2FoldChange) 
  


 
fairePeaks.osaEctopic %>% data.frame
dependent.go <- select(org.Dm.eg.db, 
       keys = test$geneId, 
       columns = c('SYMBOL','GO','ENTREZID'),
       keytype = "FLYBASE")

dependent.terms <- select(GO.db, 
       keys = dependent.go$GO, 
       columns = c('DEFINITION', 'GOID', 'TERM'))

dependent <- dplyr::bind_cols(dependent.go, dependent.terms)
#library(GOSemSim)

#dmGO <- GOSemSim::godata('org.Dm.eg.db', ont = 'MF')

dependent.entrez <- dependent %>% .$ENTREZID

ggo.dependent <- clusterProfiler::enrichGO(gene = dependent.entrez,
                        OrgDb = org.Dm.eg.db,
                         ont = 'all',
                         readable = T)
ggo.dependent %>%
  data.frame() %>%
#  dplyr::filter(grepl('sensory',Description)) %>%
  dplyr::mutate(symbol = stringr::str_split(geneID,'/')) %>%
  tidyr::unnest(symbol)
  dplyr::select(Description, qvalue, symbol) %>%
  dplyr::group_by(symbol) %>%
  dplyr::tally() %>%
  dplyr::arrange(desc(n))
  
png('rPlots/4-ectopic.png', width = 10, height = 8, res = 300, units = 'in')
enrichplot::dotplot(ggo.dependent, split = 'ONTOLOGY', font.size = 12, label_format = 80) + facet_grid(SYMBOL~., scale = 'free')
dev.off()

```

# weirdness
```{r}
brd.aBr <- read.csv('../DATA/Microscopy/aBr-quant-test.csv')

brd.aBr %>%
  dplyr::group_by(id) %>%
  dplyr::mutate(ratio = mean / mean[genotype == 'wt']) %>%
  dplyr::filter(genotype == 'kd' & driver == 'ci') %>%
  ggplot(aes(genotype, ratio)) +
  geom_boxplot() +
  geom_point(size = 4, position = position_jitter(width = 0.02), shape = 21, fill = 'grey10', alpha = 0.2) +
  ylab('Mean Grey Value Ratio (KD/WT)') +
  theme(axis.title.x = element_blank(),
        axis.text = element_text(size = 12),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
ggsave('rPlots/aBr_osaKd_quant-test.png', width = 2, height = 3)
 
```

